// @compodoc/compodoc/dist/template-playground-server.js
'use strict';

var logger = require('./logger-kbUbohEP.js');
var path = require('path');
var fs = require('fs-extra');
var http = require('http');
var crypto = require('crypto');
var os = require('os');
var child_process = require('child_process');

function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var path__namespace = /*#__PURE__*/_interopNamespaceDefault(path);
var fs__namespace = /*#__PURE__*/_interopNamespaceDefault(fs);
var http__namespace = /*#__PURE__*/_interopNamespaceDefault(http);
var crypto__namespace = /*#__PURE__*/_interopNamespaceDefault(crypto);
var os__namespace = /*#__PURE__*/_interopNamespaceDefault(os);

var polka = require('polka');
var sirv = require('sirv');
var _a = require('body-parser'), json = _a.json, urlencoded = _a.urlencoded;
var send = require('@polka/send-type');
var archiver = require('archiver');
var TemplatePlaygroundServer = /** @class */ (function () {
    function TemplatePlaygroundServer(port) {
        this.sessions = new Map();
        this.ipToSessionId = new Map();
        this.debounceTimers = new Map();
        this.signalHandlers = new Map();
        this.port = port || parseInt(process.env.PLAYGROUND_PORT || process.env.PORT || '3001', 10);
        this.app = polka();
        this.setupPaths();
        this.initializeHandlebars();
        this.setupMiddleware();
        this.setupRoutes();
        this.startSessionCleanup();
        this.setupSignalHandlers();
    }
    /**
     * Get the underlying HTTP server instance for testing purposes
     * @returns HTTP server instance or null if not started
     */
    TemplatePlaygroundServer.prototype.getHttpServer = function () {
        var _a;
        // Polka stores the actual HTTP server in the .server property
        // This is needed for Supertest compatibility which expects a Node.js HTTP server
        return ((_a = this.server) === null || _a === void 0 ? void 0 : _a.server) || null;
    };
    TemplatePlaygroundServer.prototype.setupSignalHandlers = function () {
        var _this = this;
        // Skip signal handlers entirely in test environment to prevent memory leaks
        if (process.env.NODE_ENV === 'test') {
            return;
        }
        // Handle CTRL+C (SIGINT) and other termination signals
        var signals = ['SIGINT', 'SIGTERM', 'SIGUSR2'];
        signals.forEach(function (signal) {
            var handler = function () { return logger.__awaiter(_this, void 0, void 0, function () {
                var error_1;
                return logger.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            logger.logger.info("Received ".concat(signal, ", shutting down Template Playground server gracefully..."));
                            _a.label = 1;
                        case 1:
                            _a.trys.push([1, 3, , 4]);
                            return [4 /*yield*/, this.stop()];
                        case 2:
                            _a.sent();
                            logger.logger.info('Server shutdown complete');
                            process.exit(0);
                            return [3 /*break*/, 4];
                        case 3:
                            error_1 = _a.sent();
                            logger.logger.error('Error during server shutdown:', error_1);
                            process.exit(1);
                            return [3 /*break*/, 4];
                        case 4: return [2 /*return*/];
                    }
                });
            }); };
            _this.signalHandlers.set(signal, handler);
            process.on(signal, handler);
        });
        // Handle uncaught exceptions (only if not already handled)
        if (process.listenerCount('uncaughtException') === 0) {
            var uncaughtHandler = function (error) { return logger.__awaiter(_this, void 0, void 0, function () {
                var stopError_1;
                return logger.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            logger.logger.error('Uncaught exception:', error);
                            _a.label = 1;
                        case 1:
                            _a.trys.push([1, 3, , 4]);
                            return [4 /*yield*/, this.stop()];
                        case 2:
                            _a.sent();
                            return [3 /*break*/, 4];
                        case 3:
                            stopError_1 = _a.sent();
                            logger.logger.error('Error during emergency shutdown:', stopError_1);
                            return [3 /*break*/, 4];
                        case 4:
                            process.exit(1);
                            return [2 /*return*/];
                    }
                });
            }); };
            this.signalHandlers.set('uncaughtException', uncaughtHandler);
            process.on('uncaughtException', uncaughtHandler);
        }
        // Handle unhandled promise rejections (only if not already handled)
        if (process.listenerCount('unhandledRejection') === 0) {
            var rejectionHandler = function (reason, promise) { return logger.__awaiter(_this, void 0, void 0, function () {
                var stopError_2;
                return logger.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            logger.logger.error('Unhandled rejection at:', promise, 'reason:', reason);
                            _a.label = 1;
                        case 1:
                            _a.trys.push([1, 3, , 4]);
                            return [4 /*yield*/, this.stop()];
                        case 2:
                            _a.sent();
                            return [3 /*break*/, 4];
                        case 3:
                            stopError_2 = _a.sent();
                            logger.logger.error('Error during emergency shutdown:', stopError_2);
                            return [3 /*break*/, 4];
                        case 4:
                            process.exit(1);
                            return [2 /*return*/];
                    }
                });
            }); };
            this.signalHandlers.set('unhandledRejection', rejectionHandler);
            process.on('unhandledRejection', rejectionHandler);
        }
    };
    TemplatePlaygroundServer.prototype.setupPaths = function () {
        // Try to find paths for distributed package first, then fall back to development paths
        // For playground-demo: check resources/playground-demo first, then src directory
        var distributedFakeProjectPath = path__namespace.join(__dirname, 'resources', 'playground-demo');
        var devFakeProjectPath = path__namespace.join(process.cwd(), 'src', 'playground-demo');
        if (fs__namespace.existsSync(distributedFakeProjectPath)) {
            this.fakeProjectPath = distributedFakeProjectPath;
        }
        else if (fs__namespace.existsSync(devFakeProjectPath)) {
            this.fakeProjectPath = devFakeProjectPath;
        }
        else {
            throw new Error('playground-demo directory not found. Please ensure it exists.');
        }
        // For templates: check if we're running from dist (distributed) or development
        var distributedTemplatesPath = path__namespace.join(__dirname, 'templates'); // When running from dist/, this is dist/templates
        var devTemplatesPath = path__namespace.join(process.cwd(), 'src', 'templates');
        var legacyTemplatesPath = path__namespace.join(process.cwd(), 'hbs-templates-copy');
        if (fs__namespace.existsSync(distributedTemplatesPath)) {
            this.originalTemplatesPath = distributedTemplatesPath;
        }
        else if (fs__namespace.existsSync(devTemplatesPath)) {
            this.originalTemplatesPath = devTemplatesPath;
        }
        else if (fs__namespace.existsSync(legacyTemplatesPath)) {
            // Keep legacy support for existing hbs-templates-copy
            this.originalTemplatesPath = legacyTemplatesPath;
        }
        else {
            throw new Error('Templates directory not found. Please ensure src/templates or dist/templates exists.');
        }
    };
    TemplatePlaygroundServer.prototype.getClientIP = function (req) {
        var _a;
        // Get IP address from various headers (handles proxies, load balancers, etc.)
        var forwarded = req.headers['x-forwarded-for'];
        var realIP = req.headers['x-real-ip'];
        var remoteAddr = ((_a = req.socket) === null || _a === void 0 ? void 0 : _a.remoteAddress) || 'unknown';
        var ip = (forwarded === null || forwarded === void 0 ? void 0 : forwarded.split(',')[0]) || realIP || remoteAddr || 'unknown';
        // Clean up IPv6 localhost
        if (ip === '::1' || ip === '::ffff:127.0.0.1') {
            ip = '127.0.0.1';
        }
        return ip;
    };
    TemplatePlaygroundServer.prototype.generateSessionIdFromIP = function (ip) {
        // Create a consistent hash from IP address
        return crypto__namespace.createHash('md5').update(ip + 'template-playground-salt').digest('hex');
    };
    TemplatePlaygroundServer.prototype.createOrGetSessionByIP = function (ip) {
        // Check if session already exists for this IP
        var existingSessionId = this.ipToSessionId.get(ip);
        if (existingSessionId && this.sessions.has(existingSessionId)) {
            var session_1 = this.sessions.get(existingSessionId);
            // Update last activity
            session_1.lastActivity = Date.now();
            logger.logger.info("\u267B\uFE0F  Reusing existing session for IP ".concat(ip, ": ").concat(existingSessionId));
            return session_1;
        }
        // Create new session
        var sessionId = this.generateSessionIdFromIP(ip);
        var templateDir = path__namespace.join(os__namespace.tmpdir(), "hbs-templates-copy-".concat(sessionId));
        var documentationDir = path__namespace.join(os__namespace.tmpdir(), "generated-documentation-".concat(sessionId));
        // Clean up any existing directories from previous sessions
        if (fs__namespace.existsSync(templateDir)) {
            fs__namespace.removeSync(templateDir);
        }
        if (fs__namespace.existsSync(documentationDir)) {
            fs__namespace.removeSync(documentationDir);
        }
        // Copy original templates to session directory
        fs__namespace.copySync(this.originalTemplatesPath, templateDir);
        fs__namespace.ensureDirSync(documentationDir);
        var session = {
            id: sessionId,
            templateDir: templateDir,
            documentationDir: documentationDir,
            lastActivity: Date.now(),
            config: {
                hideGenerator: false,
                disableSourceCode: false,
                disableGraph: false,
                disableCoverage: false,
                disablePrivate: false,
                disableProtected: false,
                disableInternal: false
            }
        };
        this.sessions.set(sessionId, session);
        this.ipToSessionId.set(ip, sessionId);
        logger.logger.info("\uD83C\uDD95 Created new session for IP ".concat(ip, ": ").concat(sessionId));
        // Generate initial documentation (skip in test mode to avoid template issues)
        if (process.env.NODE_ENV !== 'test') {
            this.generateDocumentation(sessionId);
        }
        return session;
    };
    TemplatePlaygroundServer.prototype.createNewSession = function (ip) {
        // Generate a unique session ID (not based on IP)
        var sessionId = crypto__namespace.randomBytes(16).toString('hex');
        var templateDir = path__namespace.join(os__namespace.tmpdir(), "hbs-templates-copy-".concat(sessionId));
        var documentationDir = path__namespace.join(os__namespace.tmpdir(), "generated-documentation-".concat(sessionId));
        // Clean up any existing directories from previous sessions
        if (fs__namespace.existsSync(templateDir)) {
            fs__namespace.removeSync(templateDir);
        }
        if (fs__namespace.existsSync(documentationDir)) {
            fs__namespace.removeSync(documentationDir);
        }
        // Copy original templates to session directory
        fs__namespace.copySync(this.originalTemplatesPath, templateDir);
        fs__namespace.ensureDirSync(documentationDir);
        var session = {
            id: sessionId,
            templateDir: templateDir,
            documentationDir: documentationDir,
            lastActivity: Date.now(),
            config: {
                hideGenerator: false,
                disableSourceCode: false,
                disableGraph: false,
                disableCoverage: false,
                disablePrivate: false,
                disableProtected: false,
                disableInternal: false,
                disableFilePath: false,
                disableOverview: false
            }
        };
        this.sessions.set(sessionId, session);
        // Don't update ipToSessionId mapping for new sessions to allow multiple sessions per IP
        logger.logger.info("\uD83C\uDD95 Created new session for IP ".concat(ip, ": ").concat(sessionId));
        // Generate initial documentation (skip in test mode to avoid template issues)
        if (process.env.NODE_ENV !== 'test') {
            this.generateDocumentation(sessionId);
        }
        return session;
    };
    TemplatePlaygroundServer.prototype.updateSessionActivity = function (sessionId) {
        var session = this.sessions.get(sessionId);
        if (session) {
            session.lastActivity = Date.now();
        }
    };
    TemplatePlaygroundServer.prototype.generateDocumentation = function (sessionId, debounce) {
        var _this = this;
        if (debounce === void 0) { debounce = false; }
        if (debounce) {
            // Clear existing timer
            var existingTimer = this.debounceTimers.get(sessionId);
            if (existingTimer) {
                clearTimeout(existingTimer);
            }
            // Set new timer for 300ms
            var timer = setTimeout(function () {
                _this.runCompoDocForSession(sessionId);
                _this.debounceTimers.delete(sessionId);
            }, 300);
            this.debounceTimers.set(sessionId, timer);
        }
        else {
            // Generate immediately
            this.runCompoDocForSession(sessionId);
        }
    };
    TemplatePlaygroundServer.prototype.runCompoDocForSession = function (sessionId) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var session, fakeProjectTsConfigPath, cliPath, cmd, config, booleanFlags, valueFlags, booleanFlags_1, booleanFlags_1_1, flag, valueFlags_1, valueFlags_1_1, flag, value, fullCmd;
            var e_1, _a, e_2, _b;
            return logger.__generator(this, function (_c) {
                session = this.sessions.get(sessionId);
                if (!session) {
                    logger.logger.error("Session ".concat(sessionId, " not found"));
                    return [2 /*return*/];
                }
                try {
                    logger.logger.info("\uD83D\uDE80 Generating documentation for session ".concat(sessionId));
                    fakeProjectTsConfigPath = path__namespace.join(this.fakeProjectPath, 'tsconfig.json');
                    cliPath = path__namespace.resolve(process.cwd(), 'bin', 'index-cli.js');
                    // In test mode, check if CLI exists before proceeding
                    if (process.env.NODE_ENV === 'test' && !fs__namespace.existsSync(cliPath)) {
                        logger.logger.warn("CLI not found in test environment: ".concat(cliPath, ". Skipping documentation generation."));
                        session.documentationGenerated = true; // Mark as generated to avoid retries
                        return [2 /*return*/];
                    }
                    cmd = [
                        "node \"".concat(cliPath, "\""),
                        "-p \"".concat(fakeProjectTsConfigPath, "\""),
                        "-d \"".concat(session.documentationDir, "\""),
                        "--templates \"".concat(session.templateDir, "\"")
                    ];
                    config = session.config || {};
                    booleanFlags = [
                        'hideGenerator', 'disableSourceCode', 'disableGraph', 'disableCoverage', 'disablePrivate', 'disableProtected', 'disableInternal',
                        'disableLifeCycleHooks', 'disableConstructors', 'disableRoutesGraph', 'disableSearch', 'disableDependencies', 'disableProperties',
                        'disableDomTree', 'disableTemplateTab', 'disableStyleTab', 'disableMainGraph', 'disableFilePath', 'disableOverview', 'hideDarkModeToggle', 'minimal', 'serve', 'open', 'watch', 'silent',
                        'coverageTest', 'coverageTestThresholdFail', 'coverageTestShowOnlyFailed'
                    ];
                    valueFlags = [
                        'theme', 'language', 'base', 'customFavicon', 'customLogo', 'assetsFolder', 'extTheme', 'includes', 'includesName', 'output', 'port', 'hostname',
                        'exportFormat', 'coverageTestThreshold', 'coverageMinimumPerFile', 'unitTestCoverage', 'gaID', 'gaSite', 'maxSearchResults', 'toggleMenuItems', 'navTabConfig'
                    ];
                    try {
                        for (booleanFlags_1 = logger.__values(booleanFlags), booleanFlags_1_1 = booleanFlags_1.next(); !booleanFlags_1_1.done; booleanFlags_1_1 = booleanFlags_1.next()) {
                            flag = booleanFlags_1_1.value;
                            if (config[flag] === true) {
                                cmd.push("--".concat(flag));
                            }
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (booleanFlags_1_1 && !booleanFlags_1_1.done && (_a = booleanFlags_1.return)) _a.call(booleanFlags_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                    try {
                        for (valueFlags_1 = logger.__values(valueFlags), valueFlags_1_1 = valueFlags_1.next(); !valueFlags_1_1.done; valueFlags_1_1 = valueFlags_1.next()) {
                            flag = valueFlags_1_1.value;
                            if (config[flag] !== undefined && config[flag] !== "") {
                                value = config[flag];
                                // For arrays/objects, stringify
                                if (Array.isArray(value) || typeof value === 'object') {
                                    value = JSON.stringify(value);
                                }
                                cmd.push("--".concat(flag, " \"").concat(value, "\""));
                            }
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (valueFlags_1_1 && !valueFlags_1_1.done && (_b = valueFlags_1.return)) _b.call(valueFlags_1);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                    fullCmd = cmd.join(' ');
                    logger.logger.info("\uD83D\uDE80 Executing CompoDoc command: ".concat(fullCmd));
                    // Log the command to a file for debugging
                    require('fs').appendFileSync('server-commands.log', "".concat(new Date().toISOString(), " - ").concat(fullCmd, "\n"));
                    // Execute with proper error handling (inherit stdio to see errors)
                    child_process.execSync(fullCmd, {
                        cwd: process.cwd(),
                        stdio: 'inherit' // Show output/errors instead of hiding them
                    });
                    this.updateSessionActivity(sessionId);
                    logger.logger.info("\u2705 Documentation generated successfully for session ".concat(sessionId));
                }
                catch (error) {
                    logger.logger.error("\u274C Error generating documentation for session ".concat(sessionId, ":"), error);
                }
                return [2 /*return*/];
            });
        });
    };
    TemplatePlaygroundServer.prototype.startSessionCleanup = function () {
        var _this = this;
        // Clean up sessions older than 1 hour every 10 minutes
        this.cleanupInterval = setInterval(function () {
            var e_3, _a;
            var cutoffTime = Date.now() - (60 * 60 * 1000); // 1 hour ago
            try {
                for (var _b = logger.__values(_this.sessions.entries()), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var _d = logger.__read(_c.value, 2), sessionId = _d[0], session = _d[1];
                    if (session.lastActivity < cutoffTime) {
                        _this.cleanupSession(sessionId);
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }, 10 * 60 * 1000); // Every 10 minutes
    };
    TemplatePlaygroundServer.prototype.cleanupSession = function (sessionId) {
        var e_4, _a;
        var session = this.sessions.get(sessionId);
        if (session) {
            try {
                // Remove directories
                if (fs__namespace.existsSync(session.templateDir)) {
                    fs__namespace.removeSync(session.templateDir);
                }
                if (fs__namespace.existsSync(session.documentationDir)) {
                    fs__namespace.removeSync(session.documentationDir);
                }
                // Clear timer if exists
                var timer = this.debounceTimers.get(sessionId);
                if (timer) {
                    clearTimeout(timer);
                    this.debounceTimers.delete(sessionId);
                }
                try {
                    // Remove IP mapping
                    for (var _b = logger.__values(this.ipToSessionId.entries()), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var _d = logger.__read(_c.value, 2), ip = _d[0], id = _d[1];
                        if (id === sessionId) {
                            this.ipToSessionId.delete(ip);
                            break;
                        }
                    }
                }
                catch (e_4_1) { e_4 = { error: e_4_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_4) throw e_4.error; }
                }
                this.sessions.delete(sessionId);
                logger.logger.info("\uD83E\uDDF9 Cleaned up session: ".concat(sessionId));
            }
            catch (error) {
                logger.logger.error("Error cleaning up session ".concat(sessionId, ":"), error);
            }
        }
    };
    TemplatePlaygroundServer.prototype.initializeHandlebars = function () {
        this.handlebars = require('handlebars');
        this.registerHandlebarsHelpers(this.handlebars, {});
    };
    TemplatePlaygroundServer.prototype.registerAvailablePartials = function () {
        return logger.__awaiter(this, void 0, void 0, function () {
            var partialsDir, partialFiles, partialFiles_1, partialFiles_1_1, file, partialName, partialPath, partialContent;
            var e_5, _a;
            return logger.__generator(this, function (_b) {
                try {
                    partialsDir = path__namespace.join(process.cwd(), 'dist/templates/partials');
                    logger.logger.info("\uD83D\uDD0D Looking for partials in: ".concat(partialsDir));
                    logger.logger.info("\uD83D\uDD0D Partials directory exists: ".concat(fs__namespace.existsSync(partialsDir)));
                    if (fs__namespace.existsSync(partialsDir)) {
                        partialFiles = fs__namespace.readdirSync(partialsDir).filter(function (file) { return file.endsWith('.hbs'); });
                        logger.logger.info("\uD83D\uDCC1 Found ".concat(partialFiles.length, " partial files: ").concat(JSON.stringify(partialFiles)));
                        try {
                            for (partialFiles_1 = logger.__values(partialFiles), partialFiles_1_1 = partialFiles_1.next(); !partialFiles_1_1.done; partialFiles_1_1 = partialFiles_1.next()) {
                                file = partialFiles_1_1.value;
                                partialName = file.replace('.hbs', '');
                                partialPath = path__namespace.join(partialsDir, file);
                                partialContent = fs__namespace.readFileSync(partialPath, 'utf8');
                                // Register the partial
                                this.handlebars.registerPartial(partialName, partialContent);
                                logger.logger.info("\u2705 Registered partial: ".concat(partialName));
                            }
                        }
                        catch (e_5_1) { e_5 = { error: e_5_1 }; }
                        finally {
                            try {
                                if (partialFiles_1_1 && !partialFiles_1_1.done && (_a = partialFiles_1.return)) _a.call(partialFiles_1);
                            }
                            finally { if (e_5) throw e_5.error; }
                        }
                    }
                    else {
                        logger.logger.warn("\u26A0\uFE0F Partials directory not found at: ".concat(partialsDir));
                    }
                }
                catch (error) {
                    logger.logger.error("\u274C Error registering partials:", error);
                }
                return [2 /*return*/];
            });
        });
    };
    TemplatePlaygroundServer.prototype.setupMiddleware = function () {
        // Add request logging for debugging
        this.app.use(function (req, res, next) {
            var headers = req.headers;
            logger.logger.info("\uD83D\uDD0D REQUEST: ".concat(req.method, " ").concat(req.url, " - User-Agent: ").concat(headers['user-agent'] || 'unknown'));
            next();
        });
        // Enable CORS for development
        this.app.use(function (req, res, next) {
            res.setHeader('Access-Control-Allow-Origin', '*');
            res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
            res.setHeader('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept, Authorization');
            if (req.method === 'OPTIONS') {
                res.statusCode = 200;
                res.end();
            }
            else {
                next();
            }
        });
        // Serve Compodoc resources at root level for relative path compatibility
        // Try dist/resources first (production), then src/resources (development/testing)
        var compodocResourcesPathDist = path__namespace.join(process.cwd(), 'dist/resources');
        var compodocResourcesPathSrc = path__namespace.join(process.cwd(), 'src/resources');
        var compodocResourcesPath = fs__namespace.existsSync(compodocResourcesPathDist) ? compodocResourcesPathDist : compodocResourcesPathSrc;
        logger.logger.info("\uD83D\uDCC1 Setting up root-level static files from: ".concat(compodocResourcesPath));
        logger.logger.info("\uD83D\uDCC1 Compodoc resources path exists: ".concat(fs__namespace.existsSync(compodocResourcesPath)));
        // Serve styles, js, images, and other resources at root level using sirv
        this.app.use('/styles', sirv(path__namespace.join(compodocResourcesPath, 'styles'), { dev: true }));
        this.app.use('/js', sirv(path__namespace.join(compodocResourcesPath, 'js'), { dev: true }));
        this.app.use('/images', sirv(path__namespace.join(compodocResourcesPath, 'images'), { dev: true }));
        this.app.use('/fonts', sirv(path__namespace.join(compodocResourcesPath, 'fonts'), { dev: true }));
        // Serve Compodoc resources under /resources path as well (for backward compatibility)
        this.app.use('/resources', sirv(compodocResourcesPath, { dev: true }));
        // Serve static files from template playground directory (index.html, app.js)
        // Try dist/resources first (production), then src/resources (development/testing)
        var playgroundStaticPathDist = path__namespace.join(process.cwd(), 'dist/resources/template-playground-app');
        var playgroundStaticPathSrc = path__namespace.join(process.cwd(), 'src/resources/template-playground-app');
        var playgroundStaticPath = fs__namespace.existsSync(playgroundStaticPathDist) ? playgroundStaticPathDist : playgroundStaticPathSrc;
        logger.logger.info("\uD83D\uDCC1 Setting up playground static files from: ".concat(playgroundStaticPath));
        logger.logger.info("\uD83D\uDCC1 Playground static path exists: ".concat(fs__namespace.existsSync(playgroundStaticPath)));
        this.app.use(sirv(playgroundStaticPath, { dev: true }));
        // Parse JSON bodies and form data using body-parser
        this.app.use(json({ limit: '10mb' }));
        this.app.use(urlencoded({ extended: true, limit: '10mb' }));
    };
    TemplatePlaygroundServer.prototype.setupRoutes = function () {
        var _this = this;
        // API route to get available templates
        this.app.get('/api/templates', this.getTemplates.bind(this));
        // API route to get template content
        this.app.get('/api/templates/:templateName', this.getTemplate.bind(this));
        // API route to get example data
        this.app.get('/api/example-data/:dataType', this.getExampleData.bind(this));
        // API route to render template with custom data
        this.app.post('/api/render', this.renderTemplate.bind(this));
        // API route to render complete page with template
        this.app.post('/api/render-page', this.renderCompletePage.bind(this));
        // API route to generate documentation with CompoDoc CLI
        this.app.post('/api/generate-docs', this.generateDocs.bind(this));
        // API route to download template package
        this.app.post('/api/download-template', this.downloadTemplatePackage.bind(this));
        // API route to download template ZIP (server-side creation)
        this.app.post('/api/session/:sessionId/download-zip', this.downloadSessionTemplateZip.bind(this));
        this.app.post('/api/session/:sessionId/download-all-templates', this.downloadAllSessionTemplates.bind(this));
        this.app.get('/api/session/:sessionId/download/all', this.downloadAllSessionTemplates.bind(this)); // Alias for compatibility
        // Session management API routes
        this.app.post('/api/session', this.createSessionAPI.bind(this));
        this.app.post('/api/session/create', this.createSessionAPI.bind(this));
        this.app.get('/api/session/:sessionId/templates', this.getSessionTemplates.bind(this));
        this.app.get('/api/session/:sessionId/template/*', this.getSessionTemplate.bind(this));
        this.app.post('/api/session/:sessionId/template/*', this.saveSessionTemplate.bind(this));
        this.app.get('/api/session/:sessionId/template-data/*', this.getSessionTemplateData.bind(this));
        this.app.post('/api/session/:sessionId/generate-docs', this.generateSessionDocs.bind(this));
        this.app.post('/api/session/:sessionId/generate', this.generateSessionDocs.bind(this)); // Alias for compatibility
        this.app.get('/api/session/:sessionId/config', this.getSessionConfig.bind(this));
        this.app.post('/api/session/:sessionId/config', this.updateSessionConfig.bind(this));
        // Serve session-specific generated documentation
        this.app.get('/api/session/:sessionId/docs/*', this.serveSessionDocs.bind(this));
        // Serve session-specific generated documentation at the expected URL pattern
        // These routes MUST come before the catch-all route
        this.app.get('/docs/:sessionId/index.html', function (req, res) {
            logger.logger.info("\uD83D\uDD0D Docs index route hit: /docs/".concat(req.params.sessionId, "/index.html"));
            var sessionId = req.params.sessionId;
            var session = _this.sessions.get(sessionId);
            if (!session) {
                logger.logger.error("\u274C Session not found: ".concat(sessionId));
                send(res, 404, { success: false, message: 'Session not found' });
                return;
            }
            _this.updateSessionActivity(sessionId);
            var fullPath = path__namespace.join(session.documentationDir, 'index.html');
            logger.logger.info("\uD83D\uDCC2 Looking for file: ".concat(fullPath));
            if (fs__namespace.existsSync(fullPath)) {
                logger.logger.info("\u2705 Serving file: ".concat(fullPath));
                var content = fs__namespace.readFileSync(fullPath);
                res.setHeader('Content-Type', 'text/html');
                res.end(content);
            }
            else {
                logger.logger.error("\u274C File not found: ".concat(fullPath));
                res.statusCode = 404;
                res.end('Documentation file not found');
            }
        });
        // Serve any file within session documentation using dynamic sirv middleware
        this.app.get('/docs/:sessionId/*', function (req, res) {
            var sessionId = req.params.sessionId;
            var session = _this.sessions.get(sessionId);
            if (!session) {
                logger.logger.error("\u274C Session not found: ".concat(sessionId));
                send(res, 404, { success: false, message: 'Session not found' });
                return;
            }
            _this.updateSessionActivity(sessionId);
            // Use sirv to serve files from the session documentation directory
            var sessionSirv = sirv(session.documentationDir, {
                dev: true,
                single: false,
                setHeaders: function (res, pathname) {
                    logger.logger.info("\u2705 Serving file via sirv: ".concat(pathname));
                }
            });
            // Remove the session prefix from the URL for sirv
            var originalUrl = req.url;
            var sessionPrefix = "/docs/".concat(sessionId);
            if (originalUrl && originalUrl.startsWith(sessionPrefix)) {
                req.url = originalUrl.substring(sessionPrefix.length) || '/';
                logger.logger.info("\uD83D\uDD0D Sirv serving: ".concat(req.url, " from ").concat(session.documentationDir));
            }
            sessionSirv(req, res, function () {
                // If sirv doesn't handle it, restore original URL and return 404
                req.url = originalUrl;
                logger.logger.error("\u274C File not found in session docs: ".concat(req.url));
                res.statusCode = 404;
                res.end('Documentation file not found');
            });
        });
        // Handle direct access to session documentation root (index.html)
        this.app.get('/docs/:sessionId', function (req, res) {
            logger.logger.info("\uD83D\uDD0D Docs root route hit: /docs/".concat(req.params.sessionId));
            var sessionId = req.params.sessionId;
            var session = _this.sessions.get(sessionId);
            if (!session) {
                logger.logger.error("\u274C Session not found: ".concat(sessionId));
                send(res, 404, { success: false, message: 'Session not found' });
                return;
            }
            _this.updateSessionActivity(sessionId);
            var fullPath = path__namespace.join(session.documentationDir, 'index.html');
            logger.logger.info("\uD83D\uDCC2 Looking for file: ".concat(fullPath));
            if (fs__namespace.existsSync(fullPath)) {
                logger.logger.info("\u2705 Serving file: ".concat(fullPath));
                var content = fs__namespace.readFileSync(fullPath);
                res.setHeader('Content-Type', 'text/html');
                res.end(content);
            }
            else {
                logger.logger.error("\u274C File not found: ".concat(fullPath));
                res.statusCode = 404;
                res.end('Documentation file not found');
            }
        });
        // Serve generated documentation files (legacy) - MUST come after session-specific routes
        // TEMPORARILY COMMENTED OUT TO TEST SESSION ROUTES
        // this.app.use('/docs', express.static(this.fakeProjectPath)); // Serve generated docs from playground-demo
        // Serve the main playground app for root path only
        this.app.get('/', function (req, res) {
            // Try dist/resources first (production), then src/resources (development/testing)
            var indexPathDist = path__namespace.join(process.cwd(), 'dist/resources/template-playground-app/index.html');
            var indexPathSrc = path__namespace.join(process.cwd(), 'src/resources/template-playground-app/index.html');
            var indexPath = fs__namespace.existsSync(indexPathDist) ? indexPathDist : indexPathSrc;
            if (fs__namespace.existsSync(indexPath)) {
                var content = fs__namespace.readFileSync(indexPath);
                res.setHeader('Content-Type', 'text/html');
                res.end(content);
            }
            else {
                res.statusCode = 404;
                res.end('Template Playground not built. Please run the build process.');
            }
        });
        // Handle any remaining non-API routes by serving the main app (for SPA routing)
        // Note: This catch-all route should be last and will handle all unmatched routes
        this.app.get('*', function (req, res) {
            // Skip API, resources, and docs routes as they are handled above
            if (req.url.startsWith('/api') || req.url.startsWith('/resources') || req.url.startsWith('/docs')) {
                res.statusCode = 404;
                res.end('Not Found');
                return;
            }
            logger.logger.warn("\u26A0\uFE0F CATCH-ALL ROUTE HIT: ".concat(req.method, " ").concat(req.url));
            // Try dist/resources first (production), then src/resources (development/testing)
            var indexPathDist = path__namespace.join(process.cwd(), 'dist/resources/template-playground-app/index.html');
            var indexPathSrc = path__namespace.join(process.cwd(), 'src/resources/template-playground-app/index.html');
            var indexPath = fs__namespace.existsSync(indexPathDist) ? indexPathDist : indexPathSrc;
            if (fs__namespace.existsSync(indexPath)) {
                var content = fs__namespace.readFileSync(indexPath);
                res.setHeader('Content-Type', 'text/html');
                res.end(content);
            }
            else {
                res.statusCode = 404;
                res.end('Template Playground not built. Please run the build process.');
            }
        });
    };
    TemplatePlaygroundServer.prototype.getTemplates = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var templatesDir_1, files, templates, error_2;
            return logger.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        templatesDir_1 = path__namespace.join(process.cwd(), 'dist/templates/partials');
                        return [4 /*yield*/, fs__namespace.readdir(templatesDir_1)];
                    case 1:
                        files = _a.sent();
                        templates = files
                            .filter(function (file) { return file.endsWith('.hbs'); })
                            .map(function (file) { return ({
                            name: file.replace('.hbs', ''),
                            filename: file,
                            path: path__namespace.join(templatesDir_1, file)
                        }); });
                        send(res, 200, templates);
                        return [3 /*break*/, 3];
                    case 2:
                        error_2 = _a.sent();
                        logger.logger.error('Error reading templates:', error_2);
                        send(res, 500, { error: 'Failed to read templates' });
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    TemplatePlaygroundServer.prototype.getTemplate = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var templateName, templatePath, content, error_3;
            return logger.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        templateName = req.params.templateName;
                        templatePath = path__namespace.join(process.cwd(), 'dist/templates/partials', "".concat(templateName, ".hbs"));
                        return [4 /*yield*/, fs__namespace.pathExists(templatePath)];
                    case 1:
                        if (!(_a.sent())) {
                            send(res, 404, { error: 'Template not found' });
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, fs__namespace.readFile(templatePath, 'utf-8')];
                    case 2:
                        content = _a.sent();
                        send(res, 200, {
                            name: templateName,
                            content: content,
                            path: templatePath
                        });
                        return [3 /*break*/, 4];
                    case 3:
                        error_3 = _a.sent();
                        logger.logger.error('Error reading template:', error_3);
                        send(res, 500, { error: 'Failed to read template' });
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    TemplatePlaygroundServer.prototype.getExampleData = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var dataType, _a, EXAMPLE_DATA, TEMPLATE_CONTEXT, wrappedData, error_4;
            var _b;
            return logger.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 2, , 3]);
                        dataType = req.params.dataType;
                        return [4 /*yield*/, Promise.resolve().then(function () { return require('./example-data-DR3xY3Rr.js'); })];
                    case 1:
                        _a = _c.sent(), EXAMPLE_DATA = _a.EXAMPLE_DATA, TEMPLATE_CONTEXT = _a.TEMPLATE_CONTEXT;
                        if (!EXAMPLE_DATA[dataType]) {
                            send(res, 404, { error: 'Example data type not found' });
                            return [2 /*return*/];
                        }
                        wrappedData = dataType === 'component' || dataType === 'directive' || dataType === 'pipe' ||
                            dataType === 'guard' || dataType === 'interceptor' || dataType === 'injectable' ||
                            dataType === 'class' || dataType === 'interface' || dataType === 'entity' ? logger.__assign((_b = {}, _b[dataType] = EXAMPLE_DATA[dataType], _b), EXAMPLE_DATA[dataType]) :
                            EXAMPLE_DATA[dataType];
                        send(res, 200, {
                            data: wrappedData,
                            context: TEMPLATE_CONTEXT
                        });
                        return [3 /*break*/, 3];
                    case 2:
                        error_4 = _c.sent();
                        logger.logger.error('Error getting example data:', error_4);
                        send(res, 500, { error: 'Failed to get example data' });
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    TemplatePlaygroundServer.prototype.renderTemplate = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var _a, templateContent, templateData, templateContext, template, rendered;
            return logger.__generator(this, function (_b) {
                try {
                    _a = req.body, templateContent = _a.templateContent, templateData = _a.templateData, templateContext = _a.templateContext;
                    if (!templateContent) {
                        send(res, 400, { error: 'Template content is required' });
                        return [2 /*return*/];
                    }
                    template = this.handlebars.compile(templateContent);
                    rendered = template(templateData || {});
                    send(res, 200, { rendered: rendered });
                }
                catch (error) {
                    logger.logger.error('Error rendering template:', error);
                    send(res, 500, {
                        error: 'Failed to render template',
                        details: error.message
                    });
                }
                return [2 /*return*/];
            });
        });
    };
    TemplatePlaygroundServer.prototype.renderCompletePage = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var _a, templateContent, templateData, templateContext, renderedContent, completePage;
            return logger.__generator(this, function (_b) {
                try {
                    _a = req.body, templateContent = _a.templateContent, templateData = _a.templateData, templateContext = _a.templateContext;
                    // Handle form data by parsing JSON strings
                    if (typeof templateData === 'string') {
                        try {
                            templateData = JSON.parse(templateData);
                        }
                        catch (e) {
                            templateData = {};
                        }
                    }
                    if (typeof templateContext === 'string') {
                        try {
                            templateContext = JSON.parse(templateContext);
                        }
                        catch (e) {
                            templateContext = {};
                        }
                    }
                    if (!templateContent) {
                        send(res, 400, { error: 'Template content is required' });
                        return [2 /*return*/];
                    }
                    renderedContent = this.generateCompodocHtml(templateData || {});
                    completePage = "<!doctype html>\n<html class=\"no-js\" lang=\"\">\n    <head>\n        <meta charset=\"utf-8\">\n        <meta http-equiv=\"x-ua-compatible\" content=\"ie=edge\">\n        <title>Template Preview - Compodoc</title>\n        <meta name=\"description\" content=\"\">\n        <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n\n        <link rel=\"icon\" type=\"image/x-icon\" href=\"/resources/images/favicon.ico\">\n        <link rel=\"stylesheet\" href=\"/resources/styles/bootstrap.min.css\">\n        <link rel=\"stylesheet\" href=\"/resources/styles/compodoc.css\">\n        <link rel=\"stylesheet\" href=\"/resources/styles/prism.css\">\n        <link rel=\"stylesheet\" href=\"/resources/styles/dark.css\">\n        <link rel=\"stylesheet\" href=\"/resources/styles/style.css\">\n    </head>\n    <body>\n        <script>\n            // Blocking script to avoid flickering dark mode\n            var useDark = window.matchMedia('(prefers-color-scheme: dark)');\n            var darkModeState = useDark.matches;\n            var darkModeStateLocal = localStorage.getItem('compodoc_darkmode-state');\n            if (darkModeStateLocal) {\n                darkModeState = darkModeStateLocal === 'true';\n            }\n            if (darkModeState) {\n                document.body.classList.add('dark');\n            }\n        </script>\n\n        <div class=\"container-fluid main\">\n            <!-- START CONTENT -->\n            <div class=\"content component\">\n                <div class=\"content-data\">\n                    ".concat(renderedContent, "\n                </div>\n            </div>\n            <!-- END CONTENT -->\n        </div>\n\n        <script>\n            var COMPODOC_CURRENT_PAGE_DEPTH = 0;\n            var COMPODOC_CURRENT_PAGE_CONTEXT = 'component';\n            var COMPODOC_CURRENT_PAGE_URL = 'component.html';\n        </script>\n\n        <script src=\"/resources/js/libs/bootstrap-native.js\"></script>\n        <script src=\"/resources/js/libs/prism.js\"></script>\n        <script src=\"/resources/js/compodoc.js\"></script>\n        <script src=\"/resources/js/tabs.js\"></script>\n        <script src=\"/resources/js/sourceCode.js\"></script>\n    </body>\n</html>");
                    res.setHeader('Content-Type', 'text/html');
                    res.end(completePage);
                }
                catch (error) {
                    logger.logger.error('Error rendering complete page:', error);
                    send(res, 500, {
                        error: 'Failed to render complete page',
                        details: error.message
                    });
                }
                return [2 /*return*/];
            });
        });
    };
    TemplatePlaygroundServer.prototype.generateDocs = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var _a, customTemplateContent, mockData, clientIP, session, sessionId, templatePath, error_5;
            return logger.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 3, , 4]);
                        _a = req.body, customTemplateContent = _a.customTemplateContent, mockData = _a.mockData;
                        // Update mock data if provided
                        if (mockData) {
                            // This part of the logic needs to be adapted to work with the new session-based system
                            // For now, we'll just log that it's not directly applicable here
                            logger.logger.warn('mockData parameter is not directly applicable in this session-based system. It will be ignored.');
                        }
                        clientIP = this.getClientIP(req);
                        session = this.createOrGetSessionByIP(clientIP);
                        sessionId = session.id;
                        if (!(customTemplateContent && req.body.templatePath)) return [3 /*break*/, 2];
                        templatePath = path__namespace.join(session.templateDir, req.body.templatePath);
                        return [4 /*yield*/, fs__namespace.writeFile(templatePath, customTemplateContent, 'utf8')];
                    case 1:
                        _b.sent();
                        _b.label = 2;
                    case 2:
                        // Generate documentation for the new session
                        this.generateDocumentation(sessionId, true); // Use debounce
                        send(res, 200, { success: true, message: 'Documentation generation initiated for a new session', sessionId: sessionId });
                        return [3 /*break*/, 4];
                    case 3:
                        error_5 = _b.sent();
                        logger.logger.error('Error generating documentation:', error_5);
                        send(res, 500, {
                            error: 'Failed to generate documentation',
                            details: error_5.message
                        });
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    TemplatePlaygroundServer.prototype.registerHandlebarsHelpers = function (Handlebars, context) {
        // Register translation helper (matches Compodoc's i18n helper pattern)
        Handlebars.registerHelper('t', function () {
            console.log('T HELPER CALLED');
            var key = arguments[0];
            var translations = {
                'components': 'Components',
                'modules': 'Modules',
                'interfaces': 'Interfaces',
                'classes': 'Classes',
                'injectables': 'Injectables',
                'pipes': 'Pipes',
                'directives': 'Directives',
                'guards': 'Guards',
                'interceptors': 'Interceptors',
                'entities': 'Entities',
                'controllers': 'Controllers',
                'info': 'Info',
                'readme': 'Readme',
                'source': 'Source',
                'template': 'Template',
                'styles': 'Styles',
                'dom-tree': 'DOM Tree',
                'file': 'File',
                'description': 'Description',
                'implements': 'Implements',
                'metadata': 'Metadata',
                'index': 'Index',
                'methods': 'Methods',
                'properties': 'Properties'
            };
            return translations[key] || key;
        });
        // Register relative URL helper
        Handlebars.registerHelper('relativeURL', function (depth) {
            var args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                args[_i - 1] = arguments[_i];
            }
            var depthValue = typeof depth === 'number' ? depth : (context.depth || 0);
            var baseUrl = '../'.repeat(depthValue);
            var pathArgs = args.slice(0, -1); // Remove Handlebars options object
            return baseUrl + pathArgs.join('/');
        });
        // Register comparison helper (matches Compodoc's CompareHelper implementation)
        Handlebars.registerHelper('compare', function () {
            var context = this;
            var a = arguments[0];
            var operator = arguments[1];
            var b = arguments[2];
            var options = arguments[3];
            if (arguments.length < 4) {
                throw new Error('handlebars Helper {{compare}} expects 4 arguments');
            }
            var result = false;
            switch (operator) {
                case 'indexof':
                    result = b.indexOf(a) !== -1;
                    break;
                case '===':
                    result = a === b;
                    break;
                case '!==':
                    result = a !== b;
                    break;
                case '>':
                    result = a > b;
                    break;
                case '<':
                    result = a < b;
                    break;
                case '>=':
                    result = a >= b;
                    break;
                case '<=':
                    result = a <= b;
                    break;
                case '==':
                    result = a == b;
                    break;
                case '!=':
                    result = a != b;
                    break;
                default:
                    throw new Error('helper {{compare}}: invalid operator: `' + operator + '`');
            }
            if (result === false) {
                return options.inverse(context);
            }
            return options.fn(context);
        });
        // Register tab helpers (matches Compodoc's IsTabEnabledHelper and IsInitialTabHelper)
        Handlebars.registerHelper('isTabEnabled', function () {
            var context = this;
            var navTabs = arguments[0];
            var tabId = arguments[1];
            var options = arguments[2];
            var isEnabled = navTabs && navTabs.some(function (tab) { return tab.id === tabId; });
            if (isEnabled) {
                return options.fn(context);
            }
            else {
                return options.inverse(context);
            }
        });
        Handlebars.registerHelper('isInitialTab', function () {
            var navTabs = arguments[0];
            var tabId = arguments[1];
            var isInitial = navTabs && navTabs.length > 0 && navTabs[0].id === tabId;
            if (isInitial) {
                return 'active in';
            }
            return '';
        });
        // Register utility helpers
        Handlebars.registerHelper('orLength', function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            var options = args.pop();
            var hasLength = args.some(function (arg) { return arg && (Array.isArray(arg) ? arg.length > 0 : arg); });
            if (hasLength) {
                return options.fn(this);
            }
            else {
                return options.inverse(this);
            }
        });
        Handlebars.registerHelper('breakComma', function (array) {
            if (Array.isArray(array)) {
                return array.join(', ');
            }
            return array;
        });
        Handlebars.registerHelper('parseDescription', function (description, depth) {
            // Simple markdown parsing - just return as HTML for now
            return new Handlebars.SafeString(description || '');
        });
        Handlebars.registerHelper('escapeSimpleQuote', function (text) {
            if (typeof text === 'string') {
                return text.replace(/'/g, "\\'");
            }
            return text;
        });
        // Register JSDoc helper
        Handlebars.registerHelper('jsdoc-code-example', function (jsdoctags, options) {
            return options.fn({ tags: jsdoctags || [] });
        });
        // Register link-type helper as a simple partial
        Handlebars.registerHelper('link-type', function (type, options) {
            if (type && type.href) {
                return new Handlebars.SafeString("<a href=\"".concat(type.href, "\" target=\"").concat(type.target || '_self', "\">").concat(type.raw || type, "</a>"));
            }
            return type;
        });
        // Register built-in block helpers
        Handlebars.registerHelper('each', Handlebars.helpers.each);
        Handlebars.registerHelper('if', Handlebars.helpers.if);
        Handlebars.registerHelper('unless', Handlebars.helpers.unless);
        Handlebars.registerHelper('with', Handlebars.helpers.with);
        // Register common partials used in templates
        Handlebars.registerPartial('component-detail', "\n            <p class=\"comment\">\n                <h3>{{t \"file\"}}</h3>\n            </p>\n            <p class=\"comment\">\n                <code>{{component.file}}</code>\n            </p>\n\n            {{#if component.description}}\n                <p class=\"comment\">\n                    <h3>{{t \"description\"}}</h3>\n                </p>\n                <p class=\"comment\">\n                    {{{parseDescription component.description depth}}}\n                </p>\n            {{/if}}\n\n            {{#if component.implements}}\n                <p class=\"comment\">\n                    <h3>{{t \"implements\"}}</h3>\n                </p>\n                <p class=\"comment\">\n                    {{#each component.implements}}\n                        <code>{{this}}</code>{{#unless @last}}, {{/unless}}\n                    {{/each}}\n                </p>\n            {{/if}}\n\n            <section data-compodoc=\"block-metadata\">\n                <h3>{{t \"metadata\"}}</h3>\n                <table class=\"table table-sm table-hover metadata\">\n                    <tbody>\n                        {{#if component.selector}}\n                        <tr>\n                            <td class=\"col-md-3\">selector</td>\n                            <td class=\"col-md-9\"><code>{{component.selector}}</code></td>\n                        </tr>\n                        {{/if}}\n                        {{#if component.templateUrl}}\n                        <tr>\n                            <td class=\"col-md-3\">templateUrl</td>\n                            <td class=\"col-md-9\"><code>{{component.templateUrl}}</code></td>\n                        </tr>\n                        {{/if}}\n                        {{#if component.styleUrls}}\n                        <tr>\n                            <td class=\"col-md-3\">styleUrls</td>\n                            <td class=\"col-md-9\"><code>{{breakComma component.styleUrls}}</code></td>\n                        </tr>\n                        {{/if}}\n                    </tbody>\n                </table>\n            </section>\n\n            {{#orLength component.properties component.methods component.inputs component.outputs}}\n                <section data-compodoc=\"block-index\">\n                    <h3 id=\"index\">{{t \"index\"}}</h3>\n                    <table class=\"table table-sm table-bordered index-table\">\n                        <tbody>\n                            {{#if component.methods}}\n                            <tr>\n                                <td class=\"col-md-4\">\n                                    <h6><b>{{t \"methods\"}}</b></h6>\n                                </td>\n                            </tr>\n                            <tr>\n                                <td class=\"col-md-4\">\n                                    <ul class=\"index-list\">\n                                        {{#each component.methods}}\n                                        <li><a href=\"#{{name}}\">{{name}}</a></li>\n                                        {{/each}}\n                                    </ul>\n                                </td>\n                            </tr>\n                            {{/if}}\n                            {{#if component.properties}}\n                            <tr>\n                                <td class=\"col-md-4\">\n                                    <h6><b>{{t \"properties\"}}</b></h6>\n                                </td>\n                            </tr>\n                            <tr>\n                                <td class=\"col-md-4\">\n                                    <ul class=\"index-list\">\n                                        {{#each component.properties}}\n                                        <li><a href=\"#{{name}}\">{{name}}</a></li>\n                                        {{/each}}\n                                    </ul>\n                                </td>\n                            </tr>\n                            {{/if}}\n                        </tbody>\n                    </table>\n                </section>\n            {{/orLength}}\n\n            {{#if component.methods}}\n                <section data-compodoc=\"block-methods\">\n                    <h3 id=\"methods\">{{t \"methods\"}}</h3>\n                    {{#each component.methods}}\n                    <table class=\"table table-sm table-bordered\">\n                        <tbody>\n                            <tr>\n                                <td class=\"col-md-4\">\n                                    <a name=\"{{name}}\"></a>\n                                    <span class=\"name\">\n                                        <span><b>{{name}}</b></span>\n                                        <a href=\"#{{name}}\"><span class=\"icon ion-ios-link\"></span></a>\n                                    </span>\n                                </td>\n                            </tr>\n                            <tr>\n                                <td class=\"col-md-4\">\n                                    <code>{{name}}({{#each args}}{{name}}: {{type}}{{#unless @last}}, {{/unless}}{{/each}})</code>\n                                </td>\n                            </tr>\n                            {{#if description}}\n                            <tr>\n                                <td class=\"col-md-4\">\n                                    <div class=\"io-description\">{{description}}</div>\n                                    <div class=\"io-description\">\n                                        <b>Returns : </b><code>{{type}}</code>\n                                    </div>\n                                </td>\n                            </tr>\n                            {{/if}}\n                        </tbody>\n                    </table>\n                    {{/each}}\n                </section>\n            {{/if}}\n        ");
        Handlebars.registerPartial('index', '<!-- Index partial placeholder -->');
        Handlebars.registerPartial('link-type', '<code>{{type}}</code>');
    };
    TemplatePlaygroundServer.prototype.generateCompodocHtml = function (data) {
        var component = data.component || {};
        var navTabs = data.navTabs || [];
        // Generate navigation tabs
        var tabsHtml = navTabs.map(function (tab, index) {
            var isActive = index === 0;
            var activeClass = isActive ? 'nav-link active' : 'nav-link';
            var labelMap = {
                'info': 'Info',
                'readme': 'Readme',
                'source': 'Source',
                'template': 'Template',
                'styles': 'Styles',
                'dom-tree': 'DOM Tree'
            };
            var label = labelMap[tab.label] || tab.label;
            return "        <li class=\"nav-item\">\n            <a href=\"".concat(tab.href, "\" class=\"").concat(activeClass, "\" role=\"tab\" id=\"").concat(tab.id, "-tab\" data-bs-toggle=\"tab\" data-link=\"").concat(tab['data-link'], "\">").concat(label, "</a>\n        </li>");
        }).join('\n');
        // Generate tab content
        var tabContentHtml = '';
        // Info tab
        if (navTabs.some(function (tab) { return tab.id === 'info'; })) {
            var isActive = navTabs[0].id === 'info';
            var activeClass = isActive ? 'active in' : '';
            tabContentHtml += "    <div class=\"tab-pane fade ".concat(activeClass, "\" id=\"info\">\n        <p class=\"comment\">\n            <h3>File</h3>\n        </p>\n        <p class=\"comment\">\n            <code>").concat(component.file || '', "</code>\n        </p>\n\n        ").concat(component.description ? "\n        <p class=\"comment\">\n            <h3>Description</h3>\n        </p>\n        <p class=\"comment\">\n            <p>".concat(component.description.replace(/\n/g, '</p>\n<p>'), "</p>\n        </p>\n        ") : '', "\n\n        ").concat(component.implements && component.implements.length > 0 ? "\n        <p class=\"comment\">\n            <h3>Implements</h3>\n        </p>\n        <p class=\"comment\">\n            ".concat(component.implements.map(function (impl) { return "<code>".concat(impl, "</code>"); }).join(', '), "\n        </p>\n        ") : '', "\n\n        <section data-compodoc=\"block-metadata\">\n            <h3>Metadata</h3>\n            <table class=\"table table-sm table-hover metadata\">\n                <tbody>\n                    ").concat(component.selector ? "\n                    <tr>\n                        <td class=\"col-md-3\">selector</td>\n                        <td class=\"col-md-9\"><code>".concat(component.selector, "</code></td>\n                    </tr>") : '', "\n                    ").concat(component.templateUrl ? "\n                    <tr>\n                        <td class=\"col-md-3\">templateUrl</td>\n                        <td class=\"col-md-9\"><code>".concat(component.templateUrl, "</code></td>\n                    </tr>") : '', "\n                    ").concat(component.styleUrls && component.styleUrls.length > 0 ? "\n                    <tr>\n                        <td class=\"col-md-3\">styleUrls</td>\n                        <td class=\"col-md-9\"><code>".concat(component.styleUrls.join(', '), "</code></td>\n                    </tr>") : '', "\n                </tbody>\n            </table>\n        </section>\n\n        ").concat(component.methods && component.methods.length > 0 ? "\n        <section data-compodoc=\"block-index\">\n            <h3 id=\"index\">Index</h3>\n            <table class=\"table table-sm table-bordered index-table\">\n                <tbody>\n                    <tr>\n                        <td class=\"col-md-4\">\n                            <h6><b>Methods</b></h6>\n                        </td>\n                    </tr>\n                    <tr>\n                        <td class=\"col-md-4\">\n                            <ul class=\"index-list\">\n                                ".concat(component.methods.map(function (method) { return "<li><a href=\"#".concat(method.name, "\">").concat(method.name, "</a></li>"); }).join('\n                                '), "\n                            </ul>\n                        </td>\n                    </tr>\n                </tbody>\n            </table>\n        </section>\n\n        <section data-compodoc=\"block-methods\">\n            <h3 id=\"methods\">Methods</h3>\n            ").concat(component.methods.map(function (method) { return "\n            <table class=\"table table-sm table-bordered\">\n                <tbody>\n                    <tr>\n                        <td class=\"col-md-4\">\n                            <a name=\"".concat(method.name, "\"></a>\n                            <span class=\"name\">\n                                <span><b>").concat(method.name, "</b></span>\n                                <a href=\"#").concat(method.name, "\"><span class=\"icon ion-ios-link\"></span></a>\n                            </span>\n                        </td>\n                    </tr>\n                    <tr>\n                        <td class=\"col-md-4\">\n                            <code>").concat(method.name, "()</code>\n                        </td>\n                    </tr>\n                    ").concat(method.description ? "\n                    <tr>\n                        <td class=\"col-md-4\">\n                            <div class=\"io-description\">".concat(method.description, "</div>\n                            <div class=\"io-description\">\n                                <b>Returns : </b><code>").concat(method.type || 'void', "</code>\n                            </div>\n                        </td>\n                    </tr>") : '', "\n                </tbody>\n            </table>"); }).join('\n            '), "\n        </section>") : '', "\n    </div>\n");
        }
        // Source tab
        if (navTabs.some(function (tab) { return tab.id === 'source'; })) {
            var isActive = navTabs[0].id === 'source';
            var activeClass = isActive ? 'active in' : '';
            tabContentHtml += "    <div class=\"tab-pane fade ".concat(activeClass, " tab-source-code\" id=\"source\">\n        <pre class=\"line-numbers compodoc-sourcecode\"><code class=\"language-typescript\">").concat(component.sourceCode || '', "</code></pre>\n    </div>\n");
        }
        // Generate complete HTML
        return "<ol class=\"breadcrumb\">\n  <li class=\"breadcrumb-item\">Components</li>\n  <li class=\"breadcrumb-item\">".concat(component.name || '', "</li>\n</ol>\n\n<ul class=\"nav nav-tabs\" role=\"tablist\">\n").concat(tabsHtml, "\n</ul>\n\n<div class=\"tab-content\">\n").concat(tabContentHtml, "</div>");
    };
    TemplatePlaygroundServer.prototype.downloadTemplatePackage = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var _a, templateType, templateContent, templateData, readme, exampleData, zipStructure;
            var _b;
            return logger.__generator(this, function (_c) {
                try {
                    _a = req.body, templateType = _a.templateType, templateContent = _a.templateContent, templateData = _a.templateData;
                    if (!templateType || !templateContent) {
                        send(res, 400, { error: 'Template type and content are required' });
                        return [2 /*return*/];
                    }
                    readme = "# Custom Compodoc Template\n\nThis package contains your customized Compodoc template for: **".concat(templateType, "**\n\n## Files Included\n\n- `templates/partials/").concat(templateType, ".hbs` - Your modified template\n- `example-data.json` - Sample data structure for testing\n- `README.md` - This file\n\n## How to Use\n\n### Option 1: Replace in existing Compodoc installation\n\n1. Backup your original template file (usually in `node_modules/@compodoc/compodoc/dist/templates/partials/").concat(templateType, ".hbs`)\n2. Replace it with the provided `").concat(templateType, ".hbs` file\n3. Regenerate your documentation with Compodoc\n\n### Option 2: Use with custom template directory\n\n1. Create a custom templates directory in your project:\n   ```\n   mkdir -p custom-templates/partials\n   ```\n\n2. Copy the `").concat(templateType, ".hbs` file to:\n   ```\n   custom-templates/partials/").concat(templateType, ".hbs\n   ```\n\n3. Run Compodoc with the custom template directory:\n   ```\n   compodoc -p tsconfig.json -d documentation --customTemplate custom-templates\n   ```\n\n## Template Variables\n\nThe template has access to these main variables:\n\n- `component` - Component information (name, description, inputs, outputs, etc.)\n- `navTabs` - Navigation tabs configuration\n- `depth` - Current page depth for relative URLs\n- `t` - Translation helper function\n\nFor a complete list of available variables, see the `example-data.json` file.\n\n## Need Help?\n\n- Compodoc Documentation: https://compodoc.app/\n- GitHub Issues: https://github.com/compodoc/compodoc/issues\n\nGenerated by Compodoc Template Playground on ").concat(new Date().toLocaleString(), "\n");
                    exampleData = {
                        template: templateType,
                        description: 'This is sample data that matches the structure used in Compodoc templates',
                        data: templateData || {}
                    };
                    zipStructure = (_b = {},
                        _b["templates/partials/".concat(templateType, ".hbs")] = templateContent,
                        _b['README.md'] = readme,
                        _b['example-data.json'] = JSON.stringify(exampleData, null, 2),
                        _b);
                    send(res, 200, {
                        success: true,
                        filename: "compodoc-".concat(templateType, "-template.zip"),
                        files: zipStructure
                    });
                }
                catch (error) {
                    logger.logger.error('Error creating template package:', error);
                    send(res, 500, {
                        error: 'Failed to create template package',
                        details: error.message
                    });
                }
                return [2 /*return*/];
            });
        });
    };
    TemplatePlaygroundServer.prototype.downloadSessionTemplateZip = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var sessionId_1, _a, templatePath_1, templateContent_1, session, templateName_1, fileName_1, error_6;
            var _this = this;
            return logger.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        sessionId_1 = req.params.sessionId;
                        _a = req.body, templatePath_1 = _a.templatePath, templateContent_1 = _a.templateContent;
                        if (!templatePath_1 || !templateContent_1) {
                            send(res, 400, { error: 'Template path and content are required' });
                            return [2 /*return*/];
                        }
                        session = this.sessions.get(sessionId_1);
                        if (!session) {
                            send(res, 404, { success: false, message: 'Session not found' });
                            return [2 /*return*/];
                        }
                        this.updateSessionActivity(sessionId_1);
                        templateName_1 = path__namespace.basename(templatePath_1, '.hbs');
                        fileName_1 = "compodoc-".concat(templateName_1, "-template.zip");
                        // Set response headers for file download
                        res.setHeader('Content-Type', 'application/zip');
                        res.setHeader('Content-Disposition', "attachment; filename=\"".concat(fileName_1, "\""));
                        // Create ZIP archive and handle it with proper promise
                        return [4 /*yield*/, new Promise(function (resolve, reject) {
                                var archive = archiver('zip', {
                                    zlib: { level: 9 } // Maximum compression
                                });
                                // Handle archive events
                                archive.on('error', function (err) {
                                    logger.logger.error('Archive error:', err);
                                    reject(new Error("Failed to create ZIP file: ".concat(err.message)));
                                });
                                archive.on('end', function () {
                                    logger.logger.info("\u2705 Template ZIP created successfully for session ".concat(sessionId_1, ": ").concat(fileName_1));
                                    resolve();
                                });
                                // Pipe archive to response
                                archive.pipe(res);
                                // Add template file to ZIP
                                archive.append(templateContent_1, { name: "templates/partials/".concat(templateName_1, ".hbs") });
                                // Create README content
                                var readme = "# Custom Compodoc Template\n\nThis package contains your customized Compodoc template for: **".concat(templateName_1, "**\n\n## Files Included\n\n- `templates/partials/").concat(templateName_1, ".hbs` - Your modified template\n- `example-data.json` - Sample data structure for testing\n- `README.md` - This file\n\n## How to Use\n\n### Option 1: Replace in existing Compodoc installation\n\n1. Backup your original template file (usually in `node_modules/@compodoc/compodoc/dist/templates/partials/").concat(templateName_1, ".hbs`)\n2. Replace it with the provided `").concat(templateName_1, ".hbs` file\n3. Regenerate your documentation with Compodoc\n\n### Option 2: Use with custom template directory\n\n1. Create a custom templates directory in your project:\n   ```\n   mkdir -p custom-templates/partials\n   ```\n\n2. Copy the `").concat(templateName_1, ".hbs` file to:\n   ```\n   custom-templates/partials/").concat(templateName_1, ".hbs\n   ```\n\n3. Run Compodoc with the custom template directory:\n   ```\n   compodoc -p tsconfig.json -d documentation --customTemplate custom-templates\n   ```\n\n## Template Variables\n\nThe template has access to these main variables:\n\n- `component` - Component information (name, description, inputs, outputs, etc.)\n- `navTabs` - Navigation tabs configuration\n- `depth` - Current page depth for relative URLs\n- `t` - Translation helper function\n\nFor a complete list of available variables, see the `example-data.json` file.\n\n## Need Help?\n\n- Compodoc Documentation: https://compodoc.app/\n- GitHub Issues: https://github.com/compodoc/compodoc/issues\n\nGenerated by Compodoc Template Playground on ").concat(new Date().toLocaleString(), "\n");
                                // Add README to ZIP
                                archive.append(readme, { name: 'README.md' });
                                // Try to get template data for the current session and template
                                _this.getSessionTemplateDataInternal(sessionId_1, templatePath_1)
                                    .then(function (templateDataResponse) {
                                    var exampleData = {
                                        template: templateName_1,
                                        description: 'This is sample data that matches the structure used in Compodoc templates',
                                        data: templateDataResponse || {}
                                    };
                                    archive.append(JSON.stringify(exampleData, null, 2), { name: 'example-data.json' });
                                })
                                    .catch(function (dataError) {
                                    logger.logger.warn('Could not get template data, using basic structure:', dataError);
                                    var basicData = {
                                        template: templateName_1,
                                        description: 'This is sample data that matches the structure used in Compodoc templates',
                                        data: { note: 'Template data could not be loaded' }
                                    };
                                    archive.append(JSON.stringify(basicData, null, 2), { name: 'example-data.json' });
                                })
                                    .finally(function () {
                                    // Finalize the archive after adding all files
                                    archive.finalize();
                                });
                            })];
                    case 1:
                        // Create ZIP archive and handle it with proper promise
                        _b.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        error_6 = _b.sent();
                        logger.logger.error('Error creating session template ZIP:', error_6);
                        if (!res.headersSent) {
                            send(res, 500, {
                                error: 'Failed to create template ZIP',
                                details: error_6.message
                            });
                        }
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    TemplatePlaygroundServer.prototype.downloadAllSessionTemplates = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var sessionId_2, session_2, fileName_2, zipBuffer, error_7;
            return logger.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        sessionId_2 = req.params.sessionId;
                        session_2 = this.sessions.get(sessionId_2);
                        if (!session_2) {
                            send(res, 404, { success: false, message: 'Session not found' });
                            return [2 /*return*/];
                        }
                        this.updateSessionActivity(sessionId_2);
                        fileName_2 = "compodoc-templates-".concat(sessionId_2, ".zip");
                        return [4 /*yield*/, new Promise(function (resolve, reject) {
                                var archive = archiver('zip', {
                                    zlib: { level: 9 } // Maximum compression
                                });
                                var chunks = [];
                                // Handle archive events
                                archive.on('error', function (err) {
                                    logger.logger.error('Archive error:', err);
                                    reject(new Error("Failed to create ZIP file: ".concat(err.message)));
                                });
                                archive.on('data', function (chunk) {
                                    chunks.push(chunk);
                                });
                                archive.on('end', function () {
                                    logger.logger.info("\u2705 All templates ZIP created successfully for session ".concat(sessionId_2, ": ").concat(fileName_2));
                                    var buffer = Buffer.concat(chunks);
                                    resolve(buffer);
                                });
                                // Add all files from the session's template directory
                                // This maintains the same structure as hbs-templates-copy-<hash>
                                archive.directory(session_2.templateDir, false);
                                // Create a comprehensive README
                                var readme = "# Compodoc Template Package\n\nThis package contains all customized Compodoc templates for session: **".concat(sessionId_2, "**\n\n## Structure\n\nThis template package has the same structure as Compodoc's default templates:\n\n- `page.hbs` - Main page template\n- `partials/` - Directory containing all partial templates:\n  - Component templates (`component.hbs`, `directive.hbs`, etc.)\n  - Block templates (`block-*.hbs`)\n  - Layout templates (`menu.hbs`, `index.hbs`, etc.)\n  - Utility templates (`search-*.hbs`, `coverage-*.hbs`, etc.)\n\n## How to Use\n\n### Option 1: Replace entire template directory\n\n1. Backup your original templates directory (usually in `node_modules/@compodoc/compodoc/dist/templates/`)\n2. Replace it with the contents of this ZIP file\n3. Regenerate your documentation with Compodoc\n\n### Option 2: Use with custom template directory\n\n1. Extract this ZIP to a directory in your project (e.g., `custom-templates/`)\n2. Run Compodoc with the custom template directory:\n   ```\n   compodoc -p tsconfig.json -d documentation --customTemplate custom-templates\n   ```\n\n### Option 3: Use specific templates only\n\n1. Extract only the templates you want to customize\n2. Place them in your custom template directory maintaining the same structure\n3. Compodoc will use your custom templates and fall back to defaults for others\n\n## Template Variables\n\nTemplates have access to comprehensive data structures including:\n\n- Component/Directive/Service information\n- Navigation and routing data\n- Documentation metadata\n- Configuration options\n- Helper functions for formatting and navigation\n\n## Need Help?\n\n- Compodoc Documentation: https://compodoc.app/\n- GitHub Issues: https://github.com/compodoc/compodoc/issues\n- Template Documentation: https://compodoc.app/guides/templates.html\n\nGenerated by Compodoc Template Playground on ").concat(new Date().toLocaleString(), "\n");
                                // Add README to ZIP root
                                archive.append(readme, { name: 'README.md' });
                                // Finalize the archive after adding all files
                                archive.finalize();
                            })];
                    case 1:
                        zipBuffer = _a.sent();
                        // Set headers and send buffer response for supertest compatibility
                        res.setHeader('Content-Type', 'application/zip');
                        res.setHeader('Content-Disposition', "attachment; filename=\"".concat(fileName_2, "\""));
                        res.setHeader('Content-Length', zipBuffer.length.toString());
                        // For testing, also add a custom header with the size
                        res.setHeader('X-Content-Size', zipBuffer.length.toString());
                        res.end(zipBuffer, 'binary');
                        return [3 /*break*/, 3];
                    case 2:
                        error_7 = _a.sent();
                        logger.logger.error('Error creating all templates ZIP:', error_7);
                        if (!res.headersSent) {
                            send(res, 500, {
                                error: 'Failed to create all templates ZIP',
                                details: error_7.message
                            });
                        }
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    TemplatePlaygroundServer.prototype.getSessionTemplateDataInternal = function (sessionId, templatePath) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var templateName, data;
            return logger.__generator(this, function (_a) {
                // Internal method to get template data without HTTP request/response
                if (!this.sessions.has(sessionId)) {
                    throw new Error('Session not found');
                }
                this.updateSessionActivity(sessionId);
                templateName = path__namespace.basename(templatePath, '.hbs');
                data = {};
                if (templateName.includes('component')) {
                    data = {
                        name: 'ExampleComponent',
                        description: 'A sample Angular component for demonstration',
                        file: 'src/app/example.component.ts',
                        selector: 'app-example',
                        templateUrl: './example.component.html',
                        styleUrls: ['./example.component.scss'],
                        inputs: [
                            { name: 'title', type: 'string', description: 'Component title' },
                            { name: 'enabled', type: 'boolean', description: 'Whether component is enabled' }
                        ],
                        outputs: [
                            { name: 'clicked', type: 'EventEmitter<void>', description: 'Emitted when clicked' }
                        ]
                    };
                }
                else if (templateName.includes('service') || templateName.includes('injectable')) {
                    data = {
                        name: 'ExampleService',
                        description: 'A sample Angular service for demonstration',
                        file: 'src/app/example.service.ts',
                        methods: [
                            { name: 'getData', returnType: 'Observable<any>', description: 'Gets data from API' },
                            { name: 'saveData', returnType: 'void', description: 'Saves data to storage' }
                        ]
                    };
                }
                else {
                    data = {
                        name: "Example".concat(templateName.charAt(0).toUpperCase() + templateName.slice(1)),
                        description: "A sample ".concat(templateName, " for demonstration"),
                        file: "src/app/example.".concat(templateName, ".ts")
                    };
                }
                return [2 /*return*/, data];
            });
        });
    };
    // Session management API methods
    TemplatePlaygroundServer.prototype.createSessionAPI = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var clientIP, forceNew, session;
            return logger.__generator(this, function (_a) {
                try {
                    clientIP = this.getClientIP(req);
                    forceNew = process.env.NODE_ENV === 'test' || req.query.forceNew === 'true';
                    session = forceNew ? this.createNewSession(clientIP) : this.createOrGetSessionByIP(clientIP);
                    send(res, 200, {
                        sessionId: session.id,
                        success: true,
                        message: 'Session created successfully',
                        ip: clientIP
                    });
                }
                catch (error) {
                    logger.logger.error('Error creating session:', error);
                    send(res, 500, {
                        success: false,
                        message: 'Failed to create session',
                        error: error instanceof Error ? error.message : 'Unknown error'
                    });
                }
                return [2 /*return*/];
            });
        });
    };
    TemplatePlaygroundServer.prototype.getSessionTemplates = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var sessionId, session, templates_1, partialsDir, mainTemplatePath, partialFiles;
            return logger.__generator(this, function (_a) {
                try {
                    sessionId = req.params.sessionId;
                    session = this.sessions.get(sessionId);
                    if (!session) {
                        send(res, 404, { success: false, message: 'Session not found' });
                        return [2 /*return*/];
                    }
                    this.updateSessionActivity(sessionId);
                    templates_1 = [];
                    partialsDir = path__namespace.join(session.templateDir, 'partials');
                    mainTemplatePath = path__namespace.join(session.templateDir, 'page.hbs');
                    if (fs__namespace.existsSync(mainTemplatePath)) {
                        templates_1.push({
                            name: 'page.hbs',
                            path: 'page.hbs',
                            type: 'template'
                        });
                    }
                    // Read partials
                    if (fs__namespace.existsSync(partialsDir)) {
                        partialFiles = fs__namespace.readdirSync(partialsDir).filter(function (file) { return file.endsWith('.hbs'); });
                        partialFiles.forEach(function (file) {
                            templates_1.push({
                                name: file,
                                path: "partials/".concat(file),
                                type: 'partial'
                            });
                        });
                    }
                    send(res, 200, { templates: templates_1, success: true });
                }
                catch (error) {
                    logger.logger.error('Error getting session templates:', error);
                    send(res, 500, {
                        success: false,
                        message: 'Failed to get templates',
                        error: error instanceof Error ? error.message : 'Unknown error'
                    });
                }
                return [2 /*return*/];
            });
        });
    };
    TemplatePlaygroundServer.prototype.getSessionTemplate = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var sessionId, templateName, urlParts, templateIndex, session, templatePath, content;
            return logger.__generator(this, function (_a) {
                try {
                    sessionId = req.params.sessionId;
                    templateName = req.params["*"];
                    urlParts = req.url.split('/');
                    templateIndex = urlParts.findIndex(function (part) { return part === 'template'; });
                    if (templateIndex !== -1 && templateIndex < urlParts.length - 1) {
                        // Get everything after 'template/' in the URL and decode it
                        templateName = decodeURIComponent(urlParts.slice(templateIndex + 1).join('/'));
                    }
                    session = this.sessions.get(sessionId);
                    // Removed debug logging - path extraction now working correctly
                    if (!session) {
                        send(res, 404, { success: false, message: 'Session not found' });
                        return [2 /*return*/];
                    }
                    this.updateSessionActivity(sessionId);
                    templatePath = path__namespace.join(session.templateDir, templateName);
                    if (!fs__namespace.existsSync(templatePath)) {
                        send(res, 404, { success: false, message: 'Template not found' });
                        return [2 /*return*/];
                    }
                    content = fs__namespace.readFileSync(templatePath, 'utf8');
                    send(res, 200, {
                        content: content,
                        success: true,
                        templateName: templateName,
                        path: templateName
                    });
                }
                catch (error) {
                    logger.logger.error('Error getting session template:', error);
                    send(res, 500, {
                        success: false,
                        message: 'Failed to get template',
                        error: error instanceof Error ? error.message : 'Unknown error'
                    });
                }
                return [2 /*return*/];
            });
        });
    };
    TemplatePlaygroundServer.prototype.saveSessionTemplate = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var sessionId, templateName, urlParts, templateIndex, content, session, templatePath;
            return logger.__generator(this, function (_a) {
                try {
                    sessionId = req.params.sessionId;
                    templateName = req.params["*"];
                    urlParts = req.url.split('/');
                    templateIndex = urlParts.findIndex(function (part) { return part === 'template'; });
                    if (templateIndex !== -1 && templateIndex < urlParts.length - 1) {
                        // Get everything after 'template/' in the URL and decode it
                        templateName = decodeURIComponent(urlParts.slice(templateIndex + 1).join('/'));
                    }
                    content = req.body.content;
                    session = this.sessions.get(sessionId);
                    // Validate required parameters
                    if (!content || typeof content !== 'string') {
                        send(res, 400, {
                            success: false,
                            message: 'Content is required and must be a string'
                        });
                        return [2 /*return*/];
                    }
                    if (!templateName) {
                        send(res, 400, {
                            success: false,
                            message: 'Template name is required'
                        });
                        return [2 /*return*/];
                    }
                    if (!session) {
                        send(res, 404, { success: false, message: 'Session not found' });
                        return [2 /*return*/];
                    }
                    this.updateSessionActivity(sessionId);
                    templatePath = path__namespace.join(session.templateDir, templateName);
                    // Ensure directory exists
                    fs__namespace.ensureDirSync(path__namespace.dirname(templatePath));
                    // Save the template content
                    fs__namespace.writeFileSync(templatePath, content, 'utf8');
                    // Trigger debounced documentation regeneration
                    this.generateDocumentation(sessionId, true);
                    send(res, 200, {
                        success: true,
                        message: 'Template saved successfully',
                        templateName: templateName
                    });
                }
                catch (error) {
                    logger.logger.error('Error saving session template:', error);
                    send(res, 500, {
                        success: false,
                        message: 'Failed to save template',
                        error: error instanceof Error ? error.message : 'Unknown error'
                    });
                }
                return [2 /*return*/];
            });
        });
    };
    TemplatePlaygroundServer.prototype.getSessionTemplateData = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var sessionId, templatePath, urlParts, templateIndex, session, templateName, compodocConfig, responseData, additionalContext, templateVariables, commonContext;
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25;
            return logger.__generator(this, function (_26) {
                try {
                    sessionId = req.params.sessionId;
                    templatePath = req.params["*"];
                    // FALLBACK: If Polka doesn't capture the full path, extract it from the URL
                    if (!templatePath || !templatePath.includes('/')) {
                        urlParts = req.url.split('/');
                        templateIndex = urlParts.findIndex(function (part) { return part === 'template-data'; });
                        if (templateIndex !== -1 && templateIndex < urlParts.length - 1) {
                            // Get everything after 'template-data/' in the URL and decode it
                            templatePath = decodeURIComponent(urlParts.slice(templateIndex + 1).join('/'));
                        }
                    }
                    if (!this.sessions.has(sessionId)) {
                        send(res, 404, {
                            success: false,
                            message: 'Session not found'
                        });
                        return [2 /*return*/];
                    }
                    this.updateSessionActivity(sessionId);
                    session = this.sessions.get(sessionId);
                    templateName = path__namespace.basename(templatePath, '.hbs');
                    compodocConfig = {
                        // Documentation Metadata
                        name: ((_a = session === null || session === void 0 ? void 0 : session.config) === null || _a === void 0 ? void 0 : _a.name) || 'Application documentation',
                        // Paths and Output
                        output: ((_b = session === null || session === void 0 ? void 0 : session.config) === null || _b === void 0 ? void 0 : _b.output) || './documentation/',
                        theme: ((_c = session === null || session === void 0 ? void 0 : session.config) === null || _c === void 0 ? void 0 : _c.theme) || 'gitbook',
                        language: ((_d = session === null || session === void 0 ? void 0 : session.config) === null || _d === void 0 ? void 0 : _d.language) || 'en-US',
                        base: ((_e = session === null || session === void 0 ? void 0 : session.config) === null || _e === void 0 ? void 0 : _e.base) || '/',
                        // Assets and Custom UI
                        customFavicon: ((_f = session === null || session === void 0 ? void 0 : session.config) === null || _f === void 0 ? void 0 : _f.customFavicon) || '',
                        customLogo: ((_g = session === null || session === void 0 ? void 0 : session.config) === null || _g === void 0 ? void 0 : _g.customLogo) || '',
                        assetsFolder: ((_h = session === null || session === void 0 ? void 0 : session.config) === null || _h === void 0 ? void 0 : _h.assetsFolder) || '',
                        extTheme: ((_j = session === null || session === void 0 ? void 0 : session.config) === null || _j === void 0 ? void 0 : _j.extTheme) || '',
                        // Feature Toggles - Disable Options
                        disableSourceCode: ((_k = session === null || session === void 0 ? void 0 : session.config) === null || _k === void 0 ? void 0 : _k.disableSourceCode) || false,
                        disableGraph: ((_l = session === null || session === void 0 ? void 0 : session.config) === null || _l === void 0 ? void 0 : _l.disableGraph) || false,
                        disableCoverage: ((_m = session === null || session === void 0 ? void 0 : session.config) === null || _m === void 0 ? void 0 : _m.disableCoverage) || false,
                        disablePrivate: ((_o = session === null || session === void 0 ? void 0 : session.config) === null || _o === void 0 ? void 0 : _o.disablePrivate) || false,
                        disableProtected: ((_p = session === null || session === void 0 ? void 0 : session.config) === null || _p === void 0 ? void 0 : _p.disableProtected) || false,
                        disableInternal: ((_q = session === null || session === void 0 ? void 0 : session.config) === null || _q === void 0 ? void 0 : _q.disableInternal) || false,
                        disableLifeCycleHooks: ((_r = session === null || session === void 0 ? void 0 : session.config) === null || _r === void 0 ? void 0 : _r.disableLifeCycleHooks) || false,
                        disableConstructors: ((_s = session === null || session === void 0 ? void 0 : session.config) === null || _s === void 0 ? void 0 : _s.disableConstructors) || false,
                        disableRoutesGraph: ((_t = session === null || session === void 0 ? void 0 : session.config) === null || _t === void 0 ? void 0 : _t.disableRoutesGraph) || false,
                        disableSearch: ((_u = session === null || session === void 0 ? void 0 : session.config) === null || _u === void 0 ? void 0 : _u.disableSearch) || false,
                        disableDependencies: ((_v = session === null || session === void 0 ? void 0 : session.config) === null || _v === void 0 ? void 0 : _v.disableDependencies) || false,
                        disableProperties: ((_w = session === null || session === void 0 ? void 0 : session.config) === null || _w === void 0 ? void 0 : _w.disableProperties) || false,
                        disableDomTree: ((_x = session === null || session === void 0 ? void 0 : session.config) === null || _x === void 0 ? void 0 : _x.disableDomTree) || false,
                        disableTemplateTab: ((_y = session === null || session === void 0 ? void 0 : session.config) === null || _y === void 0 ? void 0 : _y.disableTemplateTab) || false,
                        disableStyleTab: ((_z = session === null || session === void 0 ? void 0 : session.config) === null || _z === void 0 ? void 0 : _z.disableStyleTab) || false,
                        disableMainGraph: ((_0 = session === null || session === void 0 ? void 0 : session.config) === null || _0 === void 0 ? void 0 : _0.disableMainGraph) || false,
                        disableFilePath: ((_1 = session === null || session === void 0 ? void 0 : session.config) === null || _1 === void 0 ? void 0 : _1.disableFilePath) || false,
                        disableOverview: ((_2 = session === null || session === void 0 ? void 0 : session.config) === null || _2 === void 0 ? void 0 : _2.disableOverview) || false,
                        // UI Options
                        hideGenerator: ((_3 = session === null || session === void 0 ? void 0 : session.config) === null || _3 === void 0 ? void 0 : _3.hideGenerator) || false,
                        hideDarkModeToggle: ((_4 = session === null || session === void 0 ? void 0 : session.config) === null || _4 === void 0 ? void 0 : _4.hideDarkModeToggle) || false,
                        minimal: ((_5 = session === null || session === void 0 ? void 0 : session.config) === null || _5 === void 0 ? void 0 : _5.minimal) || false,
                        // Additional Content
                        includes: ((_6 = session === null || session === void 0 ? void 0 : session.config) === null || _6 === void 0 ? void 0 : _6.includes) || '',
                        includesName: ((_7 = session === null || session === void 0 ? void 0 : session.config) === null || _7 === void 0 ? void 0 : _7.includesName) || 'Additional documentation',
                        // Serving Options
                        port: ((_8 = session === null || session === void 0 ? void 0 : session.config) === null || _8 === void 0 ? void 0 : _8.port) || 8080,
                        hostname: ((_9 = session === null || session === void 0 ? void 0 : session.config) === null || _9 === void 0 ? void 0 : _9.hostname) || '127.0.0.1',
                        serve: ((_10 = session === null || session === void 0 ? void 0 : session.config) === null || _10 === void 0 ? void 0 : _10.serve) || false,
                        open: ((_11 = session === null || session === void 0 ? void 0 : session.config) === null || _11 === void 0 ? void 0 : _11.open) || false,
                        watch: ((_12 = session === null || session === void 0 ? void 0 : session.config) === null || _12 === void 0 ? void 0 : _12.watch) || false,
                        // Export Options
                        exportFormat: ((_13 = session === null || session === void 0 ? void 0 : session.config) === null || _13 === void 0 ? void 0 : _13.exportFormat) || 'html',
                        // Coverage Options
                        coverageTest: ((_14 = session === null || session === void 0 ? void 0 : session.config) === null || _14 === void 0 ? void 0 : _14.coverageTest) || false,
                        coverageTestThreshold: ((_15 = session === null || session === void 0 ? void 0 : session.config) === null || _15 === void 0 ? void 0 : _15.coverageTestThreshold) || 70,
                        coverageMinimumPerFile: ((_16 = session === null || session === void 0 ? void 0 : session.config) === null || _16 === void 0 ? void 0 : _16.coverageMinimumPerFile) || 0,
                        coverageTestThresholdFail: ((_17 = session === null || session === void 0 ? void 0 : session.config) === null || _17 === void 0 ? void 0 : _17.coverageTestThresholdFail) || true,
                        coverageTestShowOnlyFailed: ((_18 = session === null || session === void 0 ? void 0 : session.config) === null || _18 === void 0 ? void 0 : _18.coverageTestShowOnlyFailed) || false,
                        unitTestCoverage: ((_19 = session === null || session === void 0 ? void 0 : session.config) === null || _19 === void 0 ? void 0 : _19.unitTestCoverage) || '',
                        // Google Analytics
                        gaID: ((_20 = session === null || session === void 0 ? void 0 : session.config) === null || _20 === void 0 ? void 0 : _20.gaID) || '',
                        gaSite: ((_21 = session === null || session === void 0 ? void 0 : session.config) === null || _21 === void 0 ? void 0 : _21.gaSite) || 'auto',
                        // Advanced Options
                        silent: ((_22 = session === null || session === void 0 ? void 0 : session.config) === null || _22 === void 0 ? void 0 : _22.silent) || false,
                        maxSearchResults: ((_23 = session === null || session === void 0 ? void 0 : session.config) === null || _23 === void 0 ? void 0 : _23.maxSearchResults) || 15,
                        // Menu Configuration (as JSON string for editing)
                        toggleMenuItems: JSON.stringify(((_24 = session === null || session === void 0 ? void 0 : session.config) === null || _24 === void 0 ? void 0 : _24.toggleMenuItems) || ['all']),
                        navTabConfig: JSON.stringify(((_25 = session === null || session === void 0 ? void 0 : session.config) === null || _25 === void 0 ? void 0 : _25.navTabConfig) || [])
                    };
                    responseData = compodocConfig;
                    additionalContext = {};
                    templateVariables = void 0;
                    // Determine template type and provide comprehensive realistic data
                    if (templateName.includes('component')) {
                        templateVariables = {
                            // Core component data
                            name: 'UserProfileComponent',
                            description: 'A comprehensive user profile management component that handles user information display and editing capabilities.',
                            file: 'src/app/components/user-profile/user-profile.component.ts',
                            selector: 'app-user-profile',
                            templateUrl: './user-profile.component.html',
                            styleUrls: ['./user-profile.component.scss', './user-profile.theme.scss'],
                            encapsulation: 'ViewEncapsulation.Emulated',
                            changeDetection: 'ChangeDetectionStrategy.OnPush',
                            // Component metadata
                            type: 'component',
                            sourceCode: 'export class UserProfileComponent implements OnInit, OnDestroy { ... }',
                            rawFile: 'user-profile.component.ts',
                            // Template and styles
                            templateData: '<div class="user-profile">\\n  <h2>{{user.name}}</h2>\\n  <p>{{user.email}}</p>\\n</div>',
                            styleUrlsData: [
                                '.user-profile { padding: 20px; }\\n.user-profile h2 { color: #333; }'
                            ],
                            stylesData: [
                                ':host { display: block; margin: 10px; }'
                            ],
                            // Inputs and Outputs
                            inputs: [
                                {
                                    name: 'user',
                                    type: 'User',
                                    description: 'The user object containing profile information',
                                    decorators: ['@Input()'],
                                    optional: false,
                                    defaultValue: null
                                },
                                {
                                    name: 'editable',
                                    type: 'boolean',
                                    description: 'Whether the profile can be edited',
                                    decorators: ['@Input()'],
                                    optional: true,
                                    defaultValue: 'false'
                                },
                                {
                                    name: 'showAvatar',
                                    type: 'boolean',
                                    description: 'Controls avatar visibility',
                                    decorators: ['@Input()'],
                                    optional: true,
                                    defaultValue: 'true'
                                }
                            ],
                            outputs: [
                                {
                                    name: 'userUpdated',
                                    type: 'EventEmitter<User>',
                                    description: 'Emitted when user profile is updated',
                                    decorators: ['@Output()']
                                },
                                {
                                    name: 'avatarClicked',
                                    type: 'EventEmitter<MouseEvent>',
                                    description: 'Emitted when user clicks on avatar',
                                    decorators: ['@Output()']
                                }
                            ],
                            // Methods
                            methods: [
                                {
                                    name: 'ngOnInit',
                                    type: 'void',
                                    description: 'Angular lifecycle hook for component initialization',
                                    args: [],
                                    returnType: 'void',
                                    modifierKind: 'public'
                                },
                                {
                                    name: 'updateProfile',
                                    type: 'Promise<void>',
                                    description: 'Updates the user profile with new information',
                                    args: [
                                        { name: 'userData', type: 'Partial<User>' }
                                    ],
                                    returnType: 'Promise<void>',
                                    modifierKind: 'public'
                                },
                                {
                                    name: 'validateForm',
                                    type: 'boolean',
                                    description: 'Validates the profile form data',
                                    args: [],
                                    returnType: 'boolean',
                                    modifierKind: 'private'
                                }
                            ],
                            // Properties
                            properties: [
                                {
                                    name: 'isLoading',
                                    type: 'boolean',
                                    description: 'Indicates if component is in loading state',
                                    defaultValue: 'false',
                                    modifierKind: 'public'
                                },
                                {
                                    name: 'form',
                                    type: 'FormGroup',
                                    description: 'Reactive form for user profile editing',
                                    modifierKind: 'public'
                                }
                            ],
                            // Host listeners and bindings
                            hostListeners: [
                                {
                                    name: 'click',
                                    args: ['$event'],
                                    description: 'Handles click events on the component'
                                }
                            ],
                            hostBindings: [
                                {
                                    name: 'class.active',
                                    value: 'isActive'
                                }
                            ],
                            // Lifecycle hooks
                            implements: ['OnInit', 'OnDestroy', 'AfterViewInit'],
                            // Dependency injection
                            constructorObj: {
                                name: 'constructor',
                                description: 'Component constructor with dependency injection',
                                args: [
                                    { name: 'userService', type: 'UserService' },
                                    { name: 'router', type: 'Router' },
                                    { name: 'cd', type: 'ChangeDetectorRef' }
                                ]
                            },
                            // Angular-specific metadata
                            providers: ['UserService'],
                            viewProviders: [],
                            queries: [],
                            exportAs: 'userProfile',
                            // Documentation metadata
                            jsdoctags: [
                                {
                                    tagName: { text: 'example' },
                                    comment: '<app-user-profile [user]="currentUser" [editable]="true"></app-user-profile>'
                                }
                            ],
                            // Coverage information (if enabled)
                            coveragePercent: 85,
                            coverageCount: '17/20',
                            status: 'good'
                        };
                        additionalContext = {
                            depth: 1,
                            breadcrumbs: [
                                { name: 'Components', url: '../components.html' },
                                { name: 'UserProfileComponent', url: '#' }
                            ]
                        };
                    }
                    else if (templateName.includes('service') || templateName.includes('injectable')) {
                        templateVariables = {
                            name: 'UserService',
                            description: 'Service responsible for managing user data and authentication operations',
                            file: 'src/app/services/user.service.ts',
                            type: 'injectable',
                            // Injectable metadata
                            providedIn: 'root',
                            decorators: ['@Injectable()'],
                            // Methods
                            methods: [
                                {
                                    name: 'getUser',
                                    type: 'Observable<User>',
                                    description: 'Retrieves user data by ID',
                                    args: [{ name: 'id', type: 'string' }],
                                    returnType: 'Observable<User>',
                                    modifierKind: 'public'
                                },
                                {
                                    name: 'updateUser',
                                    type: 'Observable<User>',
                                    description: 'Updates user information',
                                    args: [
                                        { name: 'id', type: 'string' },
                                        { name: 'userData', type: 'Partial<User>' }
                                    ],
                                    returnType: 'Observable<User>',
                                    modifierKind: 'public'
                                },
                                {
                                    name: 'deleteUser',
                                    type: 'Observable<void>',
                                    description: 'Deletes a user account',
                                    args: [{ name: 'id', type: 'string' }],
                                    returnType: 'Observable<void>',
                                    modifierKind: 'public'
                                }
                            ],
                            // Properties
                            properties: [
                                {
                                    name: 'currentUser$',
                                    type: 'BehaviorSubject<User | null>',
                                    description: 'Observable stream of current user state',
                                    modifierKind: 'private'
                                },
                                {
                                    name: 'apiUrl',
                                    type: 'string',
                                    description: 'Base URL for user API endpoints',
                                    defaultValue: '"/api/users"',
                                    modifierKind: 'private'
                                }
                            ],
                            // Constructor
                            constructorObj: {
                                name: 'constructor',
                                description: 'Service constructor with HTTP client injection',
                                args: [
                                    { name: 'http', type: 'HttpClient' },
                                    { name: 'config', type: 'AppConfig' }
                                ]
                            },
                            // Coverage
                            coveragePercent: 92,
                            coverageCount: '23/25'
                        };
                    }
                    else if (templateName.includes('module')) {
                        templateVariables = {
                            name: 'UserModule',
                            description: 'Feature module containing user-related components and services',
                            file: 'src/app/modules/user/user.module.ts',
                            type: 'module',
                            // Module metadata
                            declarations: [
                                { name: 'UserProfileComponent', type: 'component' },
                                { name: 'UserListComponent', type: 'component' },
                                { name: 'UserCardDirective', type: 'directive' }
                            ],
                            imports: [
                                { name: 'CommonModule', type: 'module' },
                                { name: 'ReactiveFormsModule', type: 'module' },
                                { name: 'RouterModule', type: 'module' }
                            ],
                            exports: [
                                { name: 'UserProfileComponent', type: 'component' },
                                { name: 'UserListComponent', type: 'component' }
                            ],
                            providers: [
                                { name: 'UserService', type: 'service' },
                                { name: 'UserResolver', type: 'resolver' }
                            ],
                            bootstrap: [],
                            schemas: []
                        };
                    }
                    else if (templateName.includes('interface')) {
                        templateVariables = {
                            name: 'User',
                            description: 'Interface defining the structure of user objects',
                            file: 'src/app/interfaces/user.interface.ts',
                            type: 'interface',
                            // Interface properties
                            properties: [
                                {
                                    name: 'id',
                                    type: 'string',
                                    description: 'Unique identifier for the user',
                                    optional: false
                                },
                                {
                                    name: 'email',
                                    type: 'string',
                                    description: 'User email address',
                                    optional: false
                                },
                                {
                                    name: 'name',
                                    type: 'string',
                                    description: 'Full name of the user',
                                    optional: false
                                },
                                {
                                    name: 'avatar',
                                    type: 'string',
                                    description: 'URL to user avatar image',
                                    optional: true
                                },
                                {
                                    name: 'role',
                                    type: 'UserRole',
                                    description: 'User role permissions',
                                    optional: true
                                }
                            ],
                            // Interface methods (if any)
                            methods: [],
                            // Index signatures
                            indexSignatures: []
                        };
                    }
                    else {
                        // Generic data for other templates (directive, pipe, guard, etc.)
                        templateVariables = {
                            name: 'ExampleItem',
                            description: 'A sample item for demonstration purposes',
                            file: 'src/app/example.ts',
                            type: 'class',
                            // Basic properties that most templates would have
                            methods: [
                                {
                                    name: 'ngOnInit',
                                    type: 'void',
                                    description: 'Lifecycle hook',
                                    args: [],
                                    returnType: 'void'
                                }
                            ],
                            properties: [
                                {
                                    name: 'isActive',
                                    type: 'boolean',
                                    description: 'Active state',
                                    defaultValue: 'false'
                                }
                            ]
                        };
                    }
                    commonContext = {
                        // Navigation and UI
                        depth: additionalContext.depth || 0,
                        breadcrumbs: additionalContext.breadcrumbs || [],
                        navTabs: compodocConfig.navTabConfig,
                        // Helper functions available in templates
                        t: function (key) { return "[Translation: ".concat(key, "]"); }, // Simulates i18n function
                        relativeURL: function (url) { return url; }, // URL helper
                        // Project information
                        projectTitle: compodocConfig.documentationMainName || compodocConfig.name || 'Documentation',
                        projectDescription: compodocConfig.documentationMainDescription || 'Documentation description',
                        // Current page context
                        pageType: templateName,
                        pageName: templateVariables.name || 'Unknown',
                        // Feature flags (from config)
                        showSourceCode: !compodocConfig.disableSourceCode,
                        showGraph: !compodocConfig.disableGraph,
                        showCoverage: !compodocConfig.disableCoverage,
                        showPrivateMembers: !compodocConfig.disablePrivate,
                        showProtectedMembers: !compodocConfig.disableProtected,
                        showInternalMembers: !compodocConfig.disableInternal
                    };
                    // Return only the Compodoc configuration options
                    send(res, 200, {
                        success: true,
                        categories: {
                            compodocConfig: {
                                title: 'Compodoc Configuration Options',
                                description: 'Edit these configuration options to customize the generated documentation. Changes will automatically regenerate the documentation.',
                                data: compodocConfig
                            }
                        },
                        // Legacy format for backward compatibility
                        data: compodocConfig,
                        context: { config: compodocConfig }
                    });
                }
                catch (error) {
                    logger.logger.error('Error getting session template data:', error);
                    send(res, 500, {
                        success: false,
                        message: 'Failed to get template data',
                        error: error instanceof Error ? error.message : 'Unknown error'
                    });
                }
                return [2 /*return*/];
            });
        });
    };
    TemplatePlaygroundServer.prototype.generateSessionDocs = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var sessionId, _a, customTemplateContent, session, templatePath, error_8;
            return logger.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 3, , 4]);
                        sessionId = req.params.sessionId;
                        _a = req.body || {}, customTemplateContent = _a.customTemplateContent, _a.mockData;
                        if (!this.sessions.has(sessionId)) {
                            send(res, 404, {
                                success: false,
                                message: 'Session not found'
                            });
                            return [2 /*return*/];
                        }
                        session = this.sessions.get(sessionId);
                        this.updateSessionActivity(sessionId);
                        if (!(customTemplateContent && req.body.templatePath)) return [3 /*break*/, 2];
                        templatePath = path__namespace.join(session.templateDir, req.body.templatePath);
                        return [4 /*yield*/, fs__namespace.writeFile(templatePath, customTemplateContent, 'utf8')];
                    case 1:
                        _b.sent();
                        _b.label = 2;
                    case 2:
                        // Generate documentation for this session
                        this.generateDocumentation(sessionId, false); // No debounce for manual generation
                        send(res, 200, {
                            success: true,
                            message: 'Documentation generation started',
                            sessionId: sessionId
                        });
                        return [3 /*break*/, 4];
                    case 3:
                        error_8 = _b.sent();
                        logger.logger.error('Error generating session documentation:', error_8);
                        send(res, 500, {
                            success: false,
                            message: 'Failed to generate documentation',
                            error: error_8 instanceof Error ? error_8.message : 'Unknown error'
                        });
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    TemplatePlaygroundServer.prototype.getSessionConfig = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var sessionId, session, fullConfig;
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25;
            return logger.__generator(this, function (_26) {
                try {
                    sessionId = req.params.sessionId;
                    session = this.sessions.get(sessionId);
                    if (!session) {
                        send(res, 404, { success: false, message: 'Session not found' });
                        return [2 /*return*/];
                    }
                    this.updateSessionActivity(sessionId);
                    fullConfig = {
                        // Documentation Metadata
                        name: ((_a = session.config) === null || _a === void 0 ? void 0 : _a.name) || 'Application documentation',
                        // Paths and Output
                        output: ((_b = session.config) === null || _b === void 0 ? void 0 : _b.output) || './documentation/',
                        theme: ((_c = session.config) === null || _c === void 0 ? void 0 : _c.theme) || 'gitbook',
                        language: ((_d = session.config) === null || _d === void 0 ? void 0 : _d.language) || 'en-US',
                        base: ((_e = session.config) === null || _e === void 0 ? void 0 : _e.base) || '/',
                        // Assets and Custom UI
                        customFavicon: ((_f = session.config) === null || _f === void 0 ? void 0 : _f.customFavicon) || '',
                        customLogo: ((_g = session.config) === null || _g === void 0 ? void 0 : _g.customLogo) || '',
                        assetsFolder: ((_h = session.config) === null || _h === void 0 ? void 0 : _h.assetsFolder) || '',
                        extTheme: ((_j = session.config) === null || _j === void 0 ? void 0 : _j.extTheme) || '',
                        // Feature Toggles - Disable Options
                        disableSourceCode: !!((_k = session.config) === null || _k === void 0 ? void 0 : _k.disableSourceCode),
                        disableGraph: !!((_l = session.config) === null || _l === void 0 ? void 0 : _l.disableGraph),
                        disableCoverage: !!((_m = session.config) === null || _m === void 0 ? void 0 : _m.disableCoverage),
                        disablePrivate: !!((_o = session.config) === null || _o === void 0 ? void 0 : _o.disablePrivate),
                        disableProtected: !!((_p = session.config) === null || _p === void 0 ? void 0 : _p.disableProtected),
                        disableInternal: !!((_q = session.config) === null || _q === void 0 ? void 0 : _q.disableInternal),
                        disableLifeCycleHooks: !!((_r = session.config) === null || _r === void 0 ? void 0 : _r.disableLifeCycleHooks),
                        disableConstructors: !!((_s = session.config) === null || _s === void 0 ? void 0 : _s.disableConstructors),
                        disableRoutesGraph: !!((_t = session.config) === null || _t === void 0 ? void 0 : _t.disableRoutesGraph),
                        disableSearch: !!((_u = session.config) === null || _u === void 0 ? void 0 : _u.disableSearch),
                        disableDependencies: !!((_v = session.config) === null || _v === void 0 ? void 0 : _v.disableDependencies),
                        disableProperties: !!((_w = session.config) === null || _w === void 0 ? void 0 : _w.disableProperties),
                        disableDomTree: !!((_x = session.config) === null || _x === void 0 ? void 0 : _x.disableDomTree),
                        disableTemplateTab: !!((_y = session.config) === null || _y === void 0 ? void 0 : _y.disableTemplateTab),
                        disableStyleTab: !!((_z = session.config) === null || _z === void 0 ? void 0 : _z.disableStyleTab),
                        disableMainGraph: !!((_0 = session.config) === null || _0 === void 0 ? void 0 : _0.disableMainGraph),
                        disableFilePath: !!((_1 = session.config) === null || _1 === void 0 ? void 0 : _1.disableFilePath),
                        disableOverview: !!((_2 = session.config) === null || _2 === void 0 ? void 0 : _2.disableOverview),
                        // UI Options
                        hideGenerator: !!((_3 = session.config) === null || _3 === void 0 ? void 0 : _3.hideGenerator),
                        hideDarkModeToggle: !!((_4 = session.config) === null || _4 === void 0 ? void 0 : _4.hideDarkModeToggle),
                        minimal: !!((_5 = session.config) === null || _5 === void 0 ? void 0 : _5.minimal),
                        // Additional Content
                        includes: ((_6 = session.config) === null || _6 === void 0 ? void 0 : _6.includes) || '',
                        includesName: ((_7 = session.config) === null || _7 === void 0 ? void 0 : _7.includesName) || 'Additional documentation',
                        // Serving Options
                        port: ((_8 = session.config) === null || _8 === void 0 ? void 0 : _8.port) || 8080,
                        hostname: ((_9 = session.config) === null || _9 === void 0 ? void 0 : _9.hostname) || '127.0.0.1',
                        serve: !!((_10 = session.config) === null || _10 === void 0 ? void 0 : _10.serve),
                        open: !!((_11 = session.config) === null || _11 === void 0 ? void 0 : _11.open),
                        watch: !!((_12 = session.config) === null || _12 === void 0 ? void 0 : _12.watch),
                        // Export Options
                        exportFormat: ((_13 = session.config) === null || _13 === void 0 ? void 0 : _13.exportFormat) || 'html',
                        // Coverage Options
                        coverageTest: !!((_14 = session.config) === null || _14 === void 0 ? void 0 : _14.coverageTest),
                        coverageTestThreshold: ((_15 = session.config) === null || _15 === void 0 ? void 0 : _15.coverageTestThreshold) || 70,
                        coverageMinimumPerFile: ((_16 = session.config) === null || _16 === void 0 ? void 0 : _16.coverageMinimumPerFile) || 0,
                        coverageTestThresholdFail: !!((_17 = session.config) === null || _17 === void 0 ? void 0 : _17.coverageTestThresholdFail),
                        coverageTestShowOnlyFailed: !!((_18 = session.config) === null || _18 === void 0 ? void 0 : _18.coverageTestShowOnlyFailed),
                        unitTestCoverage: ((_19 = session.config) === null || _19 === void 0 ? void 0 : _19.unitTestCoverage) || '',
                        // Google Analytics
                        gaID: ((_20 = session.config) === null || _20 === void 0 ? void 0 : _20.gaID) || '',
                        gaSite: ((_21 = session.config) === null || _21 === void 0 ? void 0 : _21.gaSite) || 'auto',
                        // Advanced Options
                        silent: !!((_22 = session.config) === null || _22 === void 0 ? void 0 : _22.silent),
                        maxSearchResults: ((_23 = session.config) === null || _23 === void 0 ? void 0 : _23.maxSearchResults) || 15,
                        // Menu Configuration (as JSON string for editing)
                        toggleMenuItems: JSON.stringify(((_24 = session.config) === null || _24 === void 0 ? void 0 : _24.toggleMenuItems) || ['all']),
                        navTabConfig: JSON.stringify(((_25 = session.config) === null || _25 === void 0 ? void 0 : _25.navTabConfig) || [])
                    };
                    send(res, 200, {
                        config: fullConfig,
                        success: true
                    });
                }
                catch (error) {
                    logger.logger.error('Error getting session config:', error);
                    send(res, 500, {
                        success: false,
                        message: 'Failed to get config',
                        error: error instanceof Error ? error.message : 'Unknown error'
                    });
                }
                return [2 /*return*/];
            });
        });
    };
    TemplatePlaygroundServer.prototype.updateSessionConfig = function (req, res) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var sessionId, config, session;
            return logger.__generator(this, function (_a) {
                try {
                    sessionId = req.params.sessionId;
                    config = req.body.config;
                    session = this.sessions.get(sessionId);
                    if (!session) {
                        send(res, 404, { success: false, message: 'Session not found' });
                        return [2 /*return*/];
                    }
                    this.updateSessionActivity(sessionId);
                    // Update session config
                    session.config = logger.__assign(logger.__assign({}, session.config), config);
                    // Trigger debounced documentation regeneration with new config
                    this.generateDocumentation(sessionId, true);
                    send(res, 200, {
                        success: true,
                        message: 'Configuration updated successfully',
                        config: session.config
                    });
                }
                catch (error) {
                    logger.logger.error('Error updating session config:', error);
                    send(res, 500, {
                        success: false,
                        message: 'Failed to update config',
                        error: error instanceof Error ? error.message : 'Unknown error'
                    });
                }
                return [2 /*return*/];
            });
        });
    };
    TemplatePlaygroundServer.prototype.serveSessionDocs = function (req, res) {
        try {
            var sessionId = req.params.sessionId;
            var session = this.sessions.get(sessionId);
            if (!session) {
                send(res, 404, { success: false, message: 'Session not found' });
                return;
            }
            this.updateSessionActivity(sessionId);
            // Remove the session part from the URL to get the file path
            var filePath = req.url.replace(/^\/api\/session\/[^\/]+\/docs/, '');
            var fullPath = path__namespace.join(session.documentationDir, filePath || 'index.html');
            if (fs__namespace.existsSync(fullPath)) {
                var content = fs__namespace.readFileSync(fullPath);
                var ext = path__namespace.extname(fullPath).toLowerCase();
                var contentType = ext === '.html' ? 'text/html' :
                    ext === '.css' ? 'text/css' :
                        ext === '.js' ? 'application/javascript' :
                            ext === '.json' ? 'application/json' :
                                'text/plain';
                res.setHeader('Content-Type', contentType);
                res.end(content);
            }
            else {
                res.statusCode = 404;
                res.end('Documentation file not found');
            }
        }
        catch (error) {
            logger.logger.error('Error serving session docs:', error);
            res.statusCode = 500;
            res.end('Error serving documentation');
        }
    };
    TemplatePlaygroundServer.prototype.isPortAvailable = function (port) {
        return logger.__awaiter(this, void 0, void 0, function () {
            return logger.__generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve) {
                        var server = http__namespace.createServer();
                        server.listen(port, function () {
                            server.close(function () { return resolve(true); });
                        });
                        server.on('error', function () { return resolve(false); });
                    })];
            });
        });
    };
    TemplatePlaygroundServer.prototype.findAvailablePort = function (startPort) {
        return logger.__awaiter(this, void 0, void 0, function () {
            var port;
            return logger.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        port = startPort;
                        _a.label = 1;
                    case 1:
                        if (!(port < startPort + 100)) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.isPortAvailable(port)];
                    case 2:
                        if (_a.sent()) {
                            return [2 /*return*/, port];
                        }
                        port++;
                        return [3 /*break*/, 1];
                    case 3: throw new Error("No available port found in range ".concat(startPort, "-").concat(startPort + 99));
                }
            });
        });
    };
    TemplatePlaygroundServer.prototype.start = function () {
        return logger.__awaiter(this, void 0, void 0, function () {
            var originalPort, _a, error_10;
            var _this = this;
            return logger.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 6, , 7]);
                        return [4 /*yield*/, this.isPortAvailable(this.port)];
                    case 1:
                        if (!!(_b.sent())) return [3 /*break*/, 5];
                        originalPort = this.port;
                        _b.label = 2;
                    case 2:
                        _b.trys.push([2, 4, , 5]);
                        _a = this;
                        return [4 /*yield*/, this.findAvailablePort(this.port + 1)];
                    case 3:
                        _a.port = _b.sent();
                        logger.logger.warn("\u26A0\uFE0F  Port ".concat(originalPort, " is in use. Using port ").concat(this.port, " instead."));
                        return [3 /*break*/, 5];
                    case 4:
                        _b.sent();
                        throw new Error("Port ".concat(originalPort, " is in use and no alternative port could be found. Please stop the process using port ").concat(originalPort, " or specify a different port."));
                    case 5:
                        this.server = this.app.listen(this.port, function () {
                            logger.logger.info("\uD83C\uDFA8 Template Playground is running at: http://localhost:".concat(_this.port));
                            logger.logger.info(' Use this tool to customize and preview Compodoc templates');
                            logger.logger.info(' Edit templates in the left panel and see live preview on the right');
                            logger.logger.info(' Export your customized templates when ready');
                            logger.logger.info('');
                            logger.logger.info('Press Ctrl+C to stop the server');
                        });
                        // Graceful shutdown
                        process.on('SIGTERM', this.stop.bind(this));
                        process.on('SIGINT', this.stop.bind(this));
                        return [3 /*break*/, 7];
                    case 6:
                        error_10 = _b.sent();
                        logger.logger.error('Failed to start Template Playground:', error_10);
                        throw error_10;
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    TemplatePlaygroundServer.prototype.stop = function () {
        var _this = this;
        return new Promise(function (resolve) {
            var e_6, _a, e_7, _b, e_8, _c;
            var _d;
            try {
                // Remove signal handlers to prevent memory leaks
                for (var _e = logger.__values(_this.signalHandlers.entries()), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var _g = logger.__read(_f.value, 2), signal = _g[0], handler = _g[1];
                    process.removeListener(signal, handler);
                }
            }
            catch (e_6_1) { e_6 = { error: e_6_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_a = _e.return)) _a.call(_e);
                }
                finally { if (e_6) throw e_6.error; }
            }
            _this.signalHandlers.clear();
            // Clear cleanup interval
            if (_this.cleanupInterval) {
                clearInterval(_this.cleanupInterval);
                _this.cleanupInterval = null;
            }
            try {
                // Clear all debounce timers
                for (var _h = logger.__values(_this.debounceTimers.values()), _j = _h.next(); !_j.done; _j = _h.next()) {
                    var timer = _j.value;
                    clearTimeout(timer);
                }
            }
            catch (e_7_1) { e_7 = { error: e_7_1 }; }
            finally {
                try {
                    if (_j && !_j.done && (_b = _h.return)) _b.call(_h);
                }
                finally { if (e_7) throw e_7.error; }
            }
            _this.debounceTimers.clear();
            try {
                // Clean up all sessions
                for (var _k = logger.__values(_this.sessions.keys()), _l = _k.next(); !_l.done; _l = _k.next()) {
                    var sessionId = _l.value;
                    _this.cleanupSession(sessionId);
                }
            }
            catch (e_8_1) { e_8 = { error: e_8_1 }; }
            finally {
                try {
                    if (_l && !_l.done && (_c = _k.return)) _c.call(_k);
                }
                finally { if (e_8) throw e_8.error; }
            }
            // Get the actual HTTP server from Polka instance
            var httpServer = (_d = _this.server) === null || _d === void 0 ? void 0 : _d.server;
            if (httpServer && typeof httpServer.close === 'function') {
                var resolved_1 = false;
                httpServer.close(function (error) {
                    if (!resolved_1) {
                        resolved_1 = true;
                        if (error) {
                            logger.logger.warn('Error closing server:', error);
                        }
                        else {
                            logger.logger.info('Template Playground server stopped');
                        }
                        resolve();
                    }
                });
                // Force close after 1 second if it hasn't closed naturally
                setTimeout(function () {
                    if (!resolved_1) {
                        resolved_1 = true;
                        logger.logger.info('Server close timeout - resolving anyway');
                        resolve();
                    }
                }, 1000);
            }
            else {
                logger.logger.info('No HTTP server to close or close method not available');
                resolve();
            }
        });
    };
    return TemplatePlaygroundServer;
}());

exports.TemplatePlaygroundServer = TemplatePlaygroundServer;
