// @verdaccio/middleware/build/middlewares/web/web-urls.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebUrlsNamespace = exports.WebUrls = void 0;
/**
 * Enum for web urls, used on the web middleware
 */
let WebUrls = exports.WebUrls = /*#__PURE__*/function (WebUrls) {
  WebUrls["sidebar_scopped_package"] = "/sidebar/:scope/:package";
  WebUrls["sidebar_package"] = "/sidebar/:package";
  WebUrls["readme_package_scoped_version"] = "/package/readme/:scope/:package/:version?";
  WebUrls["readme_package_version"] = "/package/readme/:package/:version?";
  WebUrls["packages_all"] = "/packages";
  WebUrls["user_login"] = "/login";
  WebUrls["user_signup"] = "/signup";
  WebUrls["search"] = "/search/:anything";
  WebUrls["reset_password"] = "/reset_password";
  return WebUrls;
}({});
/**
 * Enum for web urls namespace, used on the web middleware
 */
let WebUrlsNamespace = exports.WebUrlsNamespace = /*#__PURE__*/function (WebUrlsNamespace) {
  WebUrlsNamespace["root"] = "/";
  WebUrlsNamespace["static"] = "/-/static/";
  WebUrlsNamespace["endpoints"] = "/-/verdaccio/";
  WebUrlsNamespace["web"] = "/-/web/";
  WebUrlsNamespace["data"] = "/data/";
  WebUrlsNamespace["sec"] = "/sec/";
  return WebUrlsNamespace;
}({});
