// @verdaccio/signature/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "SignOptionsSignature", {
  enumerable: true,
  get: function () {
    return _jwtToken.SignOptionsSignature;
  }
});
Object.defineProperty(exports, "VerifyOptionsSignature", {
  enumerable: true,
  get: function () {
    return _jwtToken.VerifyOptionsSignature;
  }
});
Object.defineProperty(exports, "aesDecrypt", {
  enumerable: true,
  get: function () {
    return _signature.aesDecrypt;
  }
});
Object.defineProperty(exports, "aesDecryptDeprecated", {
  enumerable: true,
  get: function () {
    return _legacySignature.aesDecryptDeprecated;
  }
});
Object.defineProperty(exports, "aesEncrypt", {
  enumerable: true,
  get: function () {
    return _signature.aesEncrypt;
  }
});
Object.defineProperty(exports, "aesEncryptDeprecated", {
  enumerable: true,
  get: function () {
    return _legacySignature.aesEncryptDeprecated;
  }
});
Object.defineProperty(exports, "generateRandomSecretKeyDeprecated", {
  enumerable: true,
  get: function () {
    return _legacySignature.generateRandomSecretKeyDeprecated;
  }
});
Object.defineProperty(exports, "parseBasicPayload", {
  enumerable: true,
  get: function () {
    return _token.parseBasicPayload;
  }
});
Object.defineProperty(exports, "signPayload", {
  enumerable: true,
  get: function () {
    return _jwtToken.signPayload;
  }
});
exports.utils = exports.types = void 0;
Object.defineProperty(exports, "verifyPayload", {
  enumerable: true,
  get: function () {
    return _jwtToken.verifyPayload;
  }
});
var _legacySignature = require("./legacy-signature");
var _signature = require("./signature");
var _jwtToken = require("./jwt-token");
var _utils = _interopRequireWildcard(require("./utils"));
exports.utils = _utils;
var _types = _interopRequireWildcard(require("./types"));
exports.types = _types;
var _token = require("./token");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
