// @verdaccio/signature/build/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "TOKEN_VALID_LENGTH", {
  enumerable: true,
  get: function () {
    return _config.TOKEN_VALID_LENGTH;
  }
});
exports.generateRandomHexString = exports.defaultTarballHashAlgorithm = exports.createTarballHash = void 0;
Object.defineProperty(exports, "generateRandomSecretKey", {
  enumerable: true,
  get: function () {
    return _config.generateRandomSecretKey;
  }
});
exports.stringToMD5 = void 0;
var _config = require("@verdaccio/config");
var _core = require("@verdaccio/core");
// @deprecated use @verdaccio/core.cryptoUtils instead
const defaultTarballHashAlgorithm = exports.defaultTarballHashAlgorithm = _core.cryptoUtils.defaultTarballHashAlgorithm;
// @deprecated use @verdaccio/core.cryptoUtils instead
const stringToMD5 = exports.stringToMD5 = _core.cryptoUtils.stringToMD5;
// @deprecated use @verdaccio/core.cryptoUtils instead
const createTarballHash = exports.createTarballHash = _core.cryptoUtils.createTarballHash;
// @deprecated use @verdaccio/core.cryptoUtils instead
const generateRandomHexString = exports.generateRandomHexString = _core.cryptoUtils.generateRandomHexString;

// @deprecated use @verdaccio/config instead
