// @verdaccio/hooks/build/notify-request.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.notifyRequest = notifyRequest;
var _debug = _interopRequireDefault(require("debug"));
var _gotCjs = _interopRequireDefault(require("got-cjs"));
var _core = require("@verdaccio/core");
var _logger = require("@verdaccio/logger");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const debug = (0, _debug.default)('verdaccio:hooks:request');
async function notifyRequest(url, options) {
  let response;
  try {
    debug('uri %o', url);
    response = _gotCjs.default.post(url, {
      body: JSON.stringify(options.body),
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      }
    });
    debug('response.status  %o', response.status);
    const body = await response.json();
    if (response.status >= _core.HTTP_STATUS.BAD_REQUEST) {
      throw new Error(body);
    }
    _logger.logger.info({
      content: options.body
    }, 'The notification @{content} has been successfully dispatched');
    return true;
  } catch (err) {
    debug('request error %o', err);
    _logger.logger.error({
      errorMessage: err?.message
    }, 'notify service has thrown an error: @{errorMessage}');
    return false;
  }
}
